// @marsulta/mailman/dist/transport/__tests__/transport.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const runtime_1 = require("../../core/runtime");
const server_1 = require("../../transport/server");
const remoteClient_1 = require("../../transport/remoteClient");
const createPacket_1 = require("../../packet/createPacket");
// Use a non-conflicting port for tests
const PORT = 17337;
(0, vitest_1.describe)("MailmanServer + RemoteMailmanClient", () => {
    let runtime;
    let server;
    let client;
    (0, vitest_1.beforeAll)(async () => {
        runtime = new runtime_1.Runtime({ dbPath: ":memory:" });
        runtime.registerRole({ name: "echo-worker", accepts: ["task.assign"] }, async (pkt) => (0, createPacket_1.createPacket)({
            taskId: pkt.taskId,
            parentPacketId: pkt.packetId,
            type: "task.result",
            sender: "echo-worker",
            target: pkt.sender,
            payload: { echoed: pkt.payload },
            status: "completed",
        }));
        runtime.registerRole({ name: "pong-role", accepts: ["health.ping"] }, async (pkt) => (0, createPacket_1.createPacket)({
            taskId: pkt.taskId,
            parentPacketId: pkt.packetId,
            type: "health.pong",
            sender: "pong-role",
            target: pkt.sender,
            payload: {},
        }));
        runtime.start();
        server = new server_1.MailmanServer(runtime, { port: PORT });
        await server.listen();
        client = new remoteClient_1.RemoteMailmanClient({
            url: `http://127.0.0.1:${PORT}`,
            timeoutMs: 5000,
        });
    });
    (0, vitest_1.afterAll)(async () => {
        await server.close();
    });
    (0, vitest_1.it)("GET /health returns running status and role names", async () => {
        const health = await client.health();
        (0, vitest_1.expect)(health.running).toBe(true);
        (0, vitest_1.expect)(health.roles).toContain("echo-worker");
        (0, vitest_1.expect)(health.roles).toContain("pong-role");
    });
    (0, vitest_1.it)("GET /roles lists all registered roles", async () => {
        const roles = await client.listRoles();
        const names = roles.map((r) => r.name);
        (0, vitest_1.expect)(names).toContain("echo-worker");
    });
    (0, vitest_1.it)("POST /packets routes a packet and returns the reply", async () => {
        const packet = (0, createPacket_1.createPacket)({
            taskId: "transport-test-1",
            type: "task.assign",
            sender: "remote-client",
            target: "echo-worker",
            payload: { hello: "world" },
        });
        const reply = await client.send(packet);
        (0, vitest_1.expect)(reply.type).toBe("task.result");
        (0, vitest_1.expect)(reply.status).toBe("completed");
        (0, vitest_1.expect)(reply.payload.echoed.hello).toBe("world");
    });
    (0, vitest_1.it)("ping returns true for a registered role", async () => {
        const alive = await client.ping("echo-worker");
        (0, vitest_1.expect)(alive).toBe(true);
    });
    (0, vitest_1.it)("ping returns false for an unknown role", async () => {
        const alive = await client.ping("nobody");
        (0, vitest_1.expect)(alive).toBe(false);
    });
    (0, vitest_1.it)("POST /packets returns error.report for unknown target", async () => {
        const packet = (0, createPacket_1.createPacket)({
            taskId: "transport-test-2",
            type: "task.assign",
            sender: "client",
            target: "ghost",
            payload: {},
        });
        const reply = await client.send(packet);
        (0, vitest_1.expect)(reply.type).toBe("error.report");
        (0, vitest_1.expect)(reply.error?.code).toBe("UNKNOWN_TARGET");
    });
});
