// @marsulta/mailman/dist/experimental/ahp-test-harness/AHPTestBus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AHPTestBus = void 0;
const ids_1 = require("../../utils/ids");
const time_1 = require("../../utils/time");
const config_1 = require("./config");
const envelopes_1 = require("./envelopes");
function createEventMessageId() {
    return (0, ids_1.newId)("maestro_msg");
}
function mergeTimeouts(overrides) {
    return {
        ...config_1.DEFAULT_AHP_TEST_TIMEOUTS,
        ...overrides,
    };
}
class AHPTestBus {
    transport;
    traces;
    subscribers = new Set();
    taskStates = new Map();
    taskSequences = new Map();
    seenMessageIds = new Set();
    registeredAgents = new Map();
    pendingSessions = new Map();
    session;
    timeouts;
    constructor(transport, traces, timeouts) {
        this.transport = transport;
        this.traces = traces;
        this.timeouts = mergeTimeouts(timeouts);
        this.transport.attachReceiver((envelope) => this.receiveEnvelope(envelope));
    }
    async connect(sessionConfig) {
        this.timeouts = mergeTimeouts(sessionConfig.timeouts ?? this.timeouts);
        if (this.session) {
            return this.session;
        }
        const sessionId = sessionConfig.sessionId ?? (0, ids_1.newId)("ahp_session");
        const taskId = `session:${sessionId}`;
        const envelope = (0, envelopes_1.createAHPTestEnvelope)({
            kind: "session.connect",
            taskId,
            correlationId: taskId,
            agentId: "workbench-test-harness",
            sessionId,
            sender: "workbench-test-harness",
            target: "ahp-test-harness",
            attempt: 1,
            payload: sessionConfig.metadata ?? {},
        });
        const responsePromise = new Promise((resolve) => {
            this.pendingSessions.set(taskId, resolve);
        });
        this.recordOutbound(envelope);
        try {
            await this.transport.sendEnvelope(envelope);
        }
        catch (error) {
            this.pendingSessions.delete(taskId);
            throw error;
        }
        const response = await this.withTimeout(responsePromise, this.timeouts.handshakeTimeoutMs, `Handshake timed out after ${this.timeouts.handshakeTimeoutMs}ms`);
        if (response.kind !== "session.connected") {
            throw new Error(`Unexpected handshake response: ${response.kind}`);
        }
        this.session = {
            sessionId,
            connectedAt: (0, time_1.now)(),
        };
        for (const agent of this.registeredAgents.values()) {
            await this.transport.registerAgent(agent, sessionId);
        }
        return this.session;
    }
    async registerAgent(agentDescriptor) {
        const session = this.requireSession();
        this.registeredAgents.set(agentDescriptor.agentId, agentDescriptor);
        await this.transport.registerAgent(agentDescriptor, session.sessionId);
    }
    async sendTask(taskEnvelope) {
        const session = this.requireSession();
        if (this.taskStates.has(taskEnvelope.taskId)) {
            return;
        }
        const state = {
            task: {
                ...taskEnvelope,
                sessionId: session.sessionId,
            },
            agentId: taskEnvelope.agentId,
            sessionId: session.sessionId,
            accepted: false,
            terminal: false,
            cancelRequested: false,
            idempotent: taskEnvelope.idempotent !== false,
            attempts: 0,
        };
        this.taskStates.set(taskEnvelope.taskId, state);
        await this.dispatchTask(state);
    }
    async cancelTask(taskId) {
        const state = this.taskStates.get(taskId);
        if (!state || state.terminal || state.cancelRequested) {
            return;
        }
        state.cancelRequested = true;
        this.clearTimer(state.ackTimer);
        const envelope = (0, envelopes_1.createAHPTestEnvelope)({
            kind: "task.cancel",
            taskId,
            correlationId: taskId,
            agentId: state.agentId,
            sessionId: state.sessionId,
            sender: "workbench-test-harness",
            target: state.agentId,
            attempt: 1,
            payload: { reason: "cancel requested" },
        });
        this.recordOutbound(envelope);
        await this.transport.sendEnvelope(envelope);
    }
    onEvent(callback) {
        this.subscribers.add(callback);
        return () => {
            this.subscribers.delete(callback);
        };
    }
    traceTask(taskId) {
        return this.traces.byTaskId(taskId);
    }
    allTraces() {
        return this.traces.all();
    }
    async restart() {
        if (!this.session) {
            return;
        }
        const previousSessionId = this.session.sessionId;
        await this.transport.disconnect(previousSessionId, [...this.registeredAgents.keys()]);
        for (const state of this.taskStates.values()) {
            this.clearTimer(state.ackTimer);
            this.clearTimer(state.inactivityTimer);
        }
        this.session = undefined;
        this.seenMessageIds.clear();
    }
    async injectEnvelope(envelope) {
        await this.transport.injectEnvelope(envelope);
    }
    recordMaestroIngress(params) {
        const messageId = createEventMessageId();
        this.traces.recordHop({
            direction: "maestro->workbench",
            messageId,
            correlationId: params.correlationId,
            taskId: params.taskId,
            agentId: params.agentId,
            sessionId: params.sessionId,
            eventType: params.eventType,
            startedAt: params.startedAt,
        });
        return messageId;
    }
    async dispatchTask(state) {
        if (state.terminal) {
            return;
        }
        state.attempts += 1;
        const envelope = (0, envelopes_1.createAHPTestEnvelope)({
            kind: "task.send",
            taskId: state.task.taskId,
            correlationId: state.task.taskId,
            agentId: state.agentId,
            sessionId: state.sessionId,
            sender: "workbench-test-harness",
            target: state.agentId,
            attempt: state.attempts,
            payload: state.task.payload,
        });
        this.recordOutbound(envelope);
        await this.transport.sendEnvelope(envelope);
        this.armAckTimer(state);
        this.armInactivityTimer(state);
    }
    receiveEnvelope(envelope) {
        this.traces.recordHop({
            direction: "ahp->workbench",
            messageId: envelope.messageId,
            correlationId: envelope.correlationId,
            taskId: envelope.taskId,
            agentId: envelope.agentId,
            sessionId: envelope.sessionId,
            eventType: envelope.kind,
            startedAt: envelope.timestamp,
        });
        if (this.seenMessageIds.has(envelope.messageId)) {
            return;
        }
        this.seenMessageIds.add(envelope.messageId);
        if (envelope.kind === "session.connected") {
            const resolve = this.pendingSessions.get(envelope.taskId);
            if (resolve) {
                this.pendingSessions.delete(envelope.taskId);
                resolve(envelope);
            }
            return;
        }
        if (envelope.kind === "agent.online") {
            this.emitEvent("AGENT_ONLINE", envelope);
            return;
        }
        if (envelope.kind === "agent.offline") {
            this.emitEvent("AGENT_OFFLINE", envelope);
            return;
        }
        const state = this.taskStates.get(envelope.taskId);
        if (!state) {
            return;
        }
        if (state.terminal) {
            return;
        }
        this.armInactivityTimer(state);
        switch (envelope.kind) {
            case "task.accept":
                state.accepted = true;
                this.clearTimer(state.ackTimer);
                this.emitEvent("TASK_ACCEPTED", envelope);
                return;
            case "task.progress":
                this.emitEvent("TASK_PROGRESS", envelope);
                return;
            case "task.partial_output":
                this.emitEvent("TASK_PARTIAL_OUTPUT", envelope);
                return;
            case "task.completed":
                state.terminal = true;
                this.finishTask(state);
                this.emitEvent("TASK_COMPLETED", envelope);
                return;
            case "task.failed":
                state.terminal = true;
                this.finishTask(state);
                this.emitEvent("TASK_FAILED", envelope);
                return;
            case "task.cancelled":
                state.terminal = true;
                this.finishTask(state);
                this.emitEvent("TASK_CANCELLED", envelope);
                return;
            default:
                return;
        }
    }
    emitEvent(eventType, source) {
        const messageId = createEventMessageId();
        const event = {
            eventType,
            messageId,
            correlationId: source.correlationId,
            taskId: source.taskId,
            agentId: source.agentId,
            sessionId: source.sessionId,
            timestamp: (0, time_1.now)(),
            sequence: this.nextSequence(source.taskId),
            payload: source.payload,
            error: source.error,
        };
        this.traces.recordHop({
            direction: "workbench->maestro",
            messageId,
            correlationId: event.correlationId,
            taskId: event.taskId,
            agentId: event.agentId,
            sessionId: event.sessionId,
            eventType: event.eventType,
            startedAt: source.timestamp,
        });
        for (const subscriber of this.subscribers) {
            subscriber(event);
        }
    }
    armAckTimer(state) {
        this.clearTimer(state.ackTimer);
        state.ackTimer = setTimeout(() => {
            if (state.accepted || state.terminal) {
                return;
            }
            if (state.idempotent && state.attempts <= this.timeouts.maxIdempotentRetries) {
                void this.dispatchTask(state);
                return;
            }
            state.terminal = true;
            this.finishTask(state);
            this.emitEvent("TASK_FAILED", {
                correlationId: state.task.taskId,
                taskId: state.task.taskId,
                agentId: state.agentId,
                sessionId: state.sessionId,
                timestamp: (0, time_1.now)(),
                error: {
                    code: "TASK_ACK_TIMEOUT",
                    message: `No TASK_ACCEPTED received within ${this.timeouts.taskAckTimeoutMs}ms`,
                },
            });
        }, this.timeouts.taskAckTimeoutMs);
    }
    armInactivityTimer(state) {
        this.clearTimer(state.inactivityTimer);
        state.inactivityTimer = setTimeout(() => {
            if (state.terminal) {
                return;
            }
            state.terminal = true;
            this.finishTask(state);
            this.emitEvent("TASK_FAILED", {
                correlationId: state.task.taskId,
                taskId: state.task.taskId,
                agentId: state.agentId,
                sessionId: state.sessionId,
                timestamp: (0, time_1.now)(),
                error: {
                    code: "TASK_INACTIVITY_TIMEOUT",
                    message: `No task activity received within ${this.timeouts.inactivityTimeoutMs}ms`,
                },
            });
        }, this.timeouts.inactivityTimeoutMs);
    }
    finishTask(state) {
        this.clearTimer(state.ackTimer);
        this.clearTimer(state.inactivityTimer);
    }
    nextSequence(taskId) {
        const next = (this.taskSequences.get(taskId) ?? 0) + 1;
        this.taskSequences.set(taskId, next);
        return next;
    }
    recordOutbound(envelope) {
        this.traces.recordHop({
            direction: "workbench->ahp",
            messageId: envelope.messageId,
            correlationId: envelope.correlationId,
            taskId: envelope.taskId,
            agentId: envelope.agentId,
            sessionId: envelope.sessionId,
            eventType: envelope.kind,
            startedAt: envelope.timestamp,
        });
    }
    requireSession() {
        if (!this.session) {
            throw new Error("AHP test harness is not connected");
        }
        return this.session;
    }
    clearTimer(timer) {
        if (timer) {
            clearTimeout(timer);
        }
    }
    withTimeout(promise, ms, message) {
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                reject(new Error(message));
            }, ms);
            promise
                .then((value) => {
                clearTimeout(timer);
                resolve(value);
            })
                .catch((error) => {
                clearTimeout(timer);
                reject(error);
            });
        });
    }
}
exports.AHPTestBus = AHPTestBus;
