// @marsulta/mailman/dist/experimental/ahp-test-harness/AHPTestAgentHost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AHPTestAgentHost = void 0;
const envelopes_1 = require("./envelopes");
class AHPTestAgentHost {
    agents = new Map();
    taskRuntimes = new Map();
    emitter;
    attachEmitter(handler) {
        this.emitter = handler;
    }
    registerAgent(agentDescriptor, sessionId) {
        this.agents.set(agentDescriptor.agentId, agentDescriptor);
        this.emitAgentStatus("agent.online", agentDescriptor.agentId, sessionId);
    }
    registeredAgentIds() {
        return [...this.agents.keys()];
    }
    emitAgentOffline(agentId, sessionId) {
        if (!this.agents.has(agentId)) {
            return;
        }
        this.emitAgentStatus("agent.offline", agentId, sessionId);
    }
    emitAgentOfflineBatch(agentIds, sessionId) {
        for (const agentId of agentIds) {
            this.emitAgentOffline(agentId, sessionId);
        }
    }
    injectEnvelope(envelope) {
        this.emitter?.(envelope);
    }
    async handleEnvelope(envelope) {
        switch (envelope.kind) {
            case "session.connect":
                this.handleSessionConnect(envelope);
                return;
            case "task.send":
                this.handleTaskSend(envelope);
                return;
            case "task.cancel":
                this.handleTaskCancel(envelope);
                return;
            default:
                return;
        }
    }
    handleSessionConnect(envelope) {
        this.injectEnvelope((0, envelopes_1.createAHPTestEnvelope)({
            kind: "session.connected",
            taskId: envelope.taskId,
            correlationId: envelope.correlationId,
            agentId: envelope.agentId,
            sessionId: envelope.sessionId,
            sender: "ahp-test-harness",
            target: envelope.sender,
            attempt: envelope.attempt,
            payload: { connected: true },
        }));
    }
    handleTaskSend(envelope) {
        const existing = this.taskRuntimes.get(envelope.taskId);
        if (existing) {
            if (existing.accepted && !existing.terminal) {
                this.emitTaskEnvelope("task.accept", envelope, { duplicateSend: true });
            }
            return;
        }
        const agent = this.agents.get(envelope.agentId);
        if (!agent) {
            this.emitTaskEnvelope("task.failed", envelope, undefined, {
                code: "UNKNOWN_AGENT",
                message: `No test agent registered for "${envelope.agentId}"`,
            });
            return;
        }
        const runtime = {
            agentId: envelope.agentId,
            accepted: false,
            terminal: false,
            cancelRequested: false,
            controller: new AbortController(),
        };
        this.taskRuntimes.set(envelope.taskId, runtime);
        const context = {
            taskId: envelope.taskId,
            sessionId: envelope.sessionId,
            agentId: envelope.agentId,
            signal: runtime.controller.signal,
            accept: (payload) => {
                runtime.accepted = true;
                return this.emitTaskEnvelope("task.accept", envelope, payload);
            },
            progress: (payload) => this.emitTaskEnvelope("task.progress", envelope, payload),
            partialOutput: (payload) => this.emitTaskEnvelope("task.partial_output", envelope, payload),
            complete: (payload) => {
                runtime.terminal = true;
                return this.emitTaskEnvelope("task.completed", envelope, payload);
            },
            fail: (error) => {
                runtime.terminal = true;
                return this.emitTaskEnvelope("task.failed", envelope, undefined, error);
            },
            cancelled: (payload) => {
                runtime.terminal = true;
                return this.emitTaskEnvelope("task.cancelled", envelope, payload);
            },
        };
        Promise.resolve(agent.handler(envelope, context))
            .then(() => {
            if (runtime.controller.signal.aborted && !runtime.terminal) {
                context.cancelled({ reason: "cancelled by orchestrator" });
            }
        })
            .catch((error) => {
            if (runtime.terminal) {
                return;
            }
            context.fail({
                code: "AGENT_HANDLER_THROWN",
                message: error instanceof Error ? error.message : "Unknown agent failure",
            });
        });
    }
    handleTaskCancel(envelope) {
        const runtime = this.taskRuntimes.get(envelope.taskId);
        if (!runtime || runtime.cancelRequested || runtime.terminal) {
            return;
        }
        runtime.cancelRequested = true;
        runtime.controller.abort();
        setTimeout(() => {
            if (!runtime.terminal) {
                runtime.terminal = true;
                this.emitTaskEnvelope("task.cancelled", envelope, {
                    reason: "cancelled by orchestrator",
                });
            }
        }, 0);
    }
    emitAgentStatus(kind, agentId, sessionId) {
        const taskId = `agent:${agentId}`;
        const envelope = (0, envelopes_1.createAHPTestEnvelope)({
            kind,
            taskId,
            correlationId: taskId,
            agentId,
            sessionId,
            sender: "ahp-test-harness",
            target: "workbench-test-harness",
            attempt: 1,
            payload: {},
        });
        this.injectEnvelope(envelope);
        return envelope;
    }
    emitTaskEnvelope(kind, outbound, payload, error) {
        const envelope = (0, envelopes_1.createAHPTestEnvelope)({
            kind,
            taskId: outbound.taskId,
            correlationId: outbound.correlationId,
            agentId: outbound.agentId,
            sessionId: outbound.sessionId,
            sender: outbound.agentId,
            target: "workbench-test-harness",
            attempt: outbound.attempt,
            payload,
            error,
        });
        this.injectEnvelope(envelope);
        return envelope;
    }
}
exports.AHPTestAgentHost = AHPTestAgentHost;
