// @marsulta/mailman/dist/experimental/ahp-test-harness/InMemoryTransport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryTransport = void 0;
const AHPTestAgentHost_1 = require("./AHPTestAgentHost");
class InMemoryTransport {
    kind = "in-memory";
    host = new AHPTestAgentHost_1.AHPTestAgentHost();
    receiver;
    constructor() {
        this.host.attachEmitter((envelope) => {
            queueMicrotask(() => {
                this.receiver?.(envelope);
            });
        });
    }
    attachReceiver(handler) {
        this.receiver = handler;
    }
    async registerAgent(agentDescriptor, sessionId) {
        this.host.registerAgent(agentDescriptor, sessionId);
    }
    async sendEnvelope(envelope) {
        await this.host.handleEnvelope(envelope);
    }
    async disconnect(sessionId, agentIds) {
        this.host.emitAgentOfflineBatch(agentIds, sessionId);
    }
    async injectEnvelope(envelope) {
        this.host.injectEnvelope(envelope);
    }
    async close() {
        return;
    }
}
exports.InMemoryTransport = InMemoryTransport;
