// @marsulta/mailman/dist/experimental/ahp-test-harness/LocalTransport.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalTransport = void 0;
const http_1 = __importDefault(require("http"));
const ws_1 = require("ws");
const AHPTestAgentHost_1 = require("./AHPTestAgentHost");
const faultInjection_1 = require("./faultInjection");
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
class LocalTransport {
    kind = "local";
    host = new AHPTestAgentHost_1.AHPTestAgentHost();
    injector;
    receiver;
    server;
    wss;
    port;
    serverSocket;
    clientSocket;
    startPromise;
    clientConnectPromise;
    constructor(options = {}, traces) {
        this.injector = new faultInjection_1.DeterministicFaultInjector(options.faultInjection, traces);
        this.host.attachEmitter((envelope) => {
            void this.sendToClient(envelope);
        });
    }
    attachReceiver(handler) {
        this.receiver = handler;
    }
    async registerAgent(agentDescriptor, sessionId) {
        await this.ensureStarted();
        this.host.registerAgent(agentDescriptor, sessionId);
    }
    async sendEnvelope(envelope) {
        await this.ensureClientConnected();
        await this.injector.transmit({ direction: "workbench->ahp", envelope }, async () => this.actualSendToServer(envelope), {
            disconnect: async () => this.closeClientSocket(),
            reconnect: async () => this.ensureClientConnected(),
        });
    }
    async disconnect(sessionId, agentIds) {
        await this.ensureStarted();
        this.host.emitAgentOfflineBatch(agentIds, sessionId);
        await sleep(10);
        await this.closeClientSocket();
    }
    async injectEnvelope(envelope) {
        await this.ensureStarted();
        this.host.injectEnvelope(envelope);
    }
    async close() {
        await this.closeClientSocket();
        if (this.wss) {
            await new Promise((resolve, reject) => {
                this.wss?.close((error) => {
                    if (error) {
                        reject(error);
                        return;
                    }
                    resolve();
                });
            });
            this.wss = undefined;
        }
        if (this.server) {
            await new Promise((resolve, reject) => {
                this.server?.close((error) => {
                    if (error) {
                        reject(error);
                        return;
                    }
                    resolve();
                });
            });
            this.server = undefined;
        }
        this.serverSocket = undefined;
        this.port = undefined;
        this.startPromise = undefined;
        this.clientConnectPromise = undefined;
    }
    async ensureStarted() {
        if (!this.startPromise) {
            this.startPromise = new Promise((resolve, reject) => {
                this.server = http_1.default.createServer();
                this.wss = new ws_1.WebSocketServer({ server: this.server });
                this.wss.on("connection", (socket) => {
                    this.serverSocket = socket;
                    socket.on("message", (raw) => {
                        this.handleServerMessage(raw.toString()).catch(reject);
                    });
                    socket.on("close", () => {
                        if (this.serverSocket === socket) {
                            this.serverSocket = undefined;
                        }
                    });
                });
                this.server.listen(0, "127.0.0.1", () => {
                    const address = this.server?.address();
                    if (!address || typeof address === "string") {
                        reject(new Error("Local transport failed to bind a TCP port"));
                        return;
                    }
                    this.port = address.port;
                    resolve();
                });
                this.server.on("error", reject);
            });
        }
        await this.startPromise;
    }
    async ensureClientConnected() {
        await this.ensureStarted();
        if (this.clientSocket?.readyState === ws_1.WebSocket.OPEN) {
            return;
        }
        if (!this.clientConnectPromise) {
            this.clientConnectPromise = (async () => {
                await this.closeClientSocket();
                await new Promise((resolve, reject) => {
                    this.clientSocket = new ws_1.WebSocket(`ws://127.0.0.1:${this.port}`);
                    this.clientSocket.on("open", () => resolve());
                    this.clientSocket.on("message", (raw) => {
                        this.handleClientMessage(raw.toString());
                    });
                    this.clientSocket.on("error", reject);
                    this.clientSocket.on("close", () => {
                        if (this.clientSocket?.readyState === ws_1.WebSocket.CLOSED) {
                            this.clientSocket = undefined;
                        }
                    });
                });
                await this.waitForServerSocket();
            })();
        }
        try {
            await this.clientConnectPromise;
        }
        finally {
            this.clientConnectPromise = undefined;
        }
    }
    async waitForServerSocket(timeoutMs = 200) {
        const started = Date.now();
        while (Date.now() - started < timeoutMs) {
            if (this.serverSocket?.readyState === ws_1.WebSocket.OPEN) {
                return;
            }
            await sleep(5);
        }
        throw new Error("Local transport server socket did not become ready");
    }
    async handleServerMessage(raw) {
        const envelope = JSON.parse(raw);
        await this.host.handleEnvelope(envelope);
    }
    handleClientMessage(raw) {
        const envelope = JSON.parse(raw);
        this.receiver?.(envelope);
    }
    async actualSendToServer(envelope) {
        await this.ensureClientConnected();
        this.clientSocket?.send(JSON.stringify(envelope));
    }
    async sendToClient(envelope) {
        await this.injector.transmit({ direction: "ahp->workbench", envelope }, async () => this.actualSendToClient(envelope), {
            disconnect: async () => this.closeClientSocket(),
            reconnect: async () => this.ensureClientConnected(),
        });
    }
    async actualSendToClient(envelope) {
        await this.ensureClientConnected();
        await this.waitForServerSocket();
        this.serverSocket?.send(JSON.stringify(envelope));
    }
    async closeClientSocket() {
        if (!this.clientSocket) {
            return;
        }
        const socket = this.clientSocket;
        await new Promise((resolve) => {
            if (socket.readyState === ws_1.WebSocket.CLOSED) {
                resolve();
                return;
            }
            socket.once("close", () => resolve());
            socket.close();
        });
        this.clientSocket = undefined;
    }
}
exports.LocalTransport = LocalTransport;
