// @marsulta/mailman/dist/experimental/ahp-test-harness/MaestroTestAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MaestroTestAdapter = void 0;
class MaestroTestAdapter {
    bus;
    sessionId;
    constructor(bus) {
        this.bus = bus;
    }
    async connect(sessionConfig) {
        const session = await this.bus.connect(sessionConfig);
        this.sessionId = session.sessionId;
        this.bus.recordMaestroIngress({
            taskId: `session:${session.sessionId}`,
            agentId: "workbench-test-harness",
            sessionId: session.sessionId,
            correlationId: `session:${session.sessionId}`,
            eventType: "SESSION_CONNECT",
        });
        return session;
    }
    async registerAgent(agentDescriptor) {
        await this.bus.registerAgent(agentDescriptor);
    }
    async sendTask(task) {
        const sessionId = this.requireSessionId();
        this.bus.recordMaestroIngress({
            taskId: task.taskId,
            agentId: task.agentId,
            sessionId,
            correlationId: task.taskId,
            eventType: "TASK_SEND",
        });
        await this.bus.sendTask({
            taskId: task.taskId,
            agentId: task.agentId,
            sessionId,
            idempotent: task.idempotent,
            payload: {
                prompt: task.prompt,
                ...(task.input ?? {}),
            },
        });
    }
    async cancelTask(taskId) {
        const sessionId = this.requireSessionId();
        this.bus.recordMaestroIngress({
            taskId,
            agentId: "workbench-test-harness",
            sessionId,
            correlationId: taskId,
            eventType: "TASK_CANCEL",
        });
        await this.bus.cancelTask(taskId);
    }
    onEvent(callback) {
        return this.bus.onEvent(callback);
    }
    traceTask(taskId) {
        return this.bus.traceTask(taskId);
    }
    allTraces() {
        return this.bus.allTraces();
    }
    requireSessionId() {
        if (!this.sessionId) {
            throw new Error("Maestro test adapter is not connected");
        }
        return this.sessionId;
    }
}
exports.MaestroTestAdapter = MaestroTestAdapter;
