// @marsulta/mailman/dist/experimental/ahp-test-harness/WorkbenchTestHarness.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkbenchTestHarness = void 0;
exports.createAHPTestHarness = createAHPTestHarness;
const AHPTestBus_1 = require("./AHPTestBus");
const config_1 = require("./config");
const InMemoryTransport_1 = require("./InMemoryTransport");
const LocalTransport_1 = require("./LocalTransport");
const logging_1 = require("./logging");
const MaestroTestAdapter_1 = require("./MaestroTestAdapter");
function createTransport(kind, traces, options) {
    return kind === "local"
        ? new LocalTransport_1.LocalTransport({ faultInjection: options.faultInjection }, traces)
        : new InMemoryTransport_1.InMemoryTransport();
}
class WorkbenchTestHarness {
    traces;
    bus;
    maestro;
    transport;
    transportKind;
    constructor(options = {}) {
        this.transportKind = options.transport ?? "in-memory";
        this.traces = new logging_1.AHPTestTraceStore();
        this.transport = createTransport(this.transportKind, this.traces, options);
        this.bus = new AHPTestBus_1.AHPTestBus(this.transport, this.traces, options.timeouts);
        this.maestro = new MaestroTestAdapter_1.MaestroTestAdapter(this.bus);
    }
    traceTask(taskId) {
        return this.traces.byTaskId(taskId);
    }
    allTraces() {
        return this.traces.all();
    }
    async injectEnvelope(envelope) {
        await this.bus.injectEnvelope(envelope);
    }
    async restart() {
        await this.bus.restart();
    }
    async close() {
        await this.transport.close();
    }
}
exports.WorkbenchTestHarness = WorkbenchTestHarness;
function createAHPTestHarness(options = {}) {
    const enabled = options.enabled ?? (0, config_1.isAHPTestModeEnabled)(options.env);
    if (!enabled) {
        return null;
    }
    return new WorkbenchTestHarness({
        timeouts: options.timeouts,
        transport: options.transport,
        faultInjection: options.faultInjection,
    });
}
