// @marsulta/mailman/dist/client/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MailmanClient = void 0;
exports.createClient = createClient;
// ─────────────────────────────────────────────
//  MailmanClient — fluent wrapper around Runtime
// ─────────────────────────────────────────────
class MailmanClient {
    runtime;
    constructor(runtime) {
        this.runtime = runtime;
    }
    // ── Role management ───────────────────────
    registerRole(role, handler) {
        this.runtime.registerRole(role, handler);
        return this;
    }
    registerStreamRole(role, handler) {
        this.runtime.registerStreamRole(role, handler);
        return this;
    }
    unregisterRole(name) {
        this.runtime.unregisterRole(name);
        return this;
    }
    listRoles() {
        return this.runtime.listRoles();
    }
    // ── Item 1: Custom packet types ───────────
    registerPacketType(type) {
        this.runtime.registerPacketType(type);
        return this;
    }
    // ── Item 2: Request/reply messaging ───────
    async send(packet) {
        return this.runtime.send(packet);
    }
    async ping(roleName) {
        return this.runtime.ping(roleName);
    }
    // ── Item 2: Fire-and-forget ───────────────
    dispatch(packet) {
        this.runtime.dispatch(packet);
        return this;
    }
    // ── Item 2: Pub/Sub ───────────────────────
    /**
     * Subscribe to packets by topic pattern.
     * Supports exact type, "task.*", or "*" for all.
     * Returns an unsubscribe function.
     */
    subscribe(topic, handler) {
        return this.runtime.subscribe(topic, handler);
    }
    publish(packet) {
        this.runtime.publish(packet);
        return this;
    }
    // ── Item 4: Streaming ─────────────────────
    async openStream(packet) {
        return this.runtime.openStream(packet);
    }
    // ── Item 6: Capability routing ────────────
    async sendToCapability(capability, packet, strategy) {
        return this.runtime.sendToCapability(capability, packet, strategy);
    }
    // ── Item 7: Ack/nack ──────────────────────
    async sendWithAck(packet) {
        return this.runtime.sendWithAck(packet);
    }
    deliverResult(taskId, result) {
        this.runtime.deliverResult(taskId, result);
        return this;
    }
    // ── Item 8: Scheduling ────────────────────
    schedulePacket(packet, deliverAt) {
        return this.runtime.schedulePacket(packet, deliverAt);
    }
    scheduleAfter(packet, delayMs) {
        return this.runtime.scheduleAfter(packet, delayMs);
    }
    scheduleRecurring(packet, intervalMs, startAt) {
        return this.runtime.scheduleRecurring(packet, intervalMs, startAt);
    }
    cancelScheduled(scheduleId) {
        return this.runtime.cancelScheduled(scheduleId);
    }
    listScheduled() {
        return this.runtime.listScheduled();
    }
    // ── Trace ─────────────────────────────────
    getTrace(packetId) {
        return this.runtime.getTrace(packetId);
    }
    getTaskTrace(taskId) {
        return this.runtime.getTaskTrace(taskId);
    }
    // ── Lifecycle ─────────────────────────────
    start() {
        this.runtime.start();
        return this;
    }
    stop() {
        this.runtime.stop();
        return this;
    }
    isRunning() {
        return this.runtime.isRunning();
    }
}
exports.MailmanClient = MailmanClient;
// ─────────────────────────────────────────────
//  Factory
// ─────────────────────────────────────────────
function createClient(runtime) {
    return new MailmanClient(runtime);
}
