// @marsulta/mailman/dist/transport/remoteClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteMailmanClient = void 0;
const http_1 = __importDefault(require("http"));
const https_1 = __importDefault(require("https"));
// ─────────────────────────────────────────────
//  RemoteMailmanClient
//
//  Sends packets to a MailmanServer over HTTP.
//  Implements the same send/ping/listRoles API
//  as MailmanClient so they're interchangeable.
// ─────────────────────────────────────────────
class RemoteMailmanClient {
    baseUrl;
    timeoutMs;
    constructor(config) {
        this.baseUrl = config.url.replace(/\/$/, ""); // strip trailing slash
        this.timeoutMs = config.timeoutMs ?? 30_000;
    }
    // ── Messaging ─────────────────────────────
    async send(packet) {
        const reply = await this.post("/packets", packet);
        return reply;
    }
    async ping(roleName) {
        try {
            const health = await this.get("/health");
            return health.roles.includes(roleName);
        }
        catch {
            return false;
        }
    }
    // ── Role info ─────────────────────────────
    async listRoles() {
        return this.get("/roles");
    }
    async health() {
        return this.get("/health");
    }
    // ── Internal HTTP helpers ─────────────────
    post(path, body) {
        return this.request("POST", path, JSON.stringify(body));
    }
    get(path) {
        return this.request("GET", path);
    }
    request(method, path, body) {
        const url = new URL(this.baseUrl + path);
        const lib = url.protocol === "https:" ? https_1.default : http_1.default;
        return new Promise((resolve, reject) => {
            const options = {
                hostname: url.hostname,
                port: url.port || (url.protocol === "https:" ? 443 : 80),
                path: url.pathname + url.search,
                method,
                headers: {
                    "Content-Type": "application/json",
                    Accept: "application/json",
                    ...(body
                        ? { "Content-Length": Buffer.byteLength(body) }
                        : {}),
                },
            };
            const req = lib.request(options, (res) => {
                const chunks = [];
                res.on("data", (chunk) => chunks.push(chunk));
                res.on("end", () => {
                    try {
                        const text = Buffer.concat(chunks).toString("utf8");
                        const parsed = JSON.parse(text);
                        const status = res.statusCode ?? 200;
                        if (status >= 400) {
                            reject(new Error(`MailmanServer returned ${status}: ${JSON.stringify(parsed)}`));
                        }
                        else {
                            resolve(parsed);
                        }
                    }
                    catch (err) {
                        reject(err);
                    }
                });
                res.on("error", reject);
            });
            req.on("error", reject);
            // Timeout
            req.setTimeout(this.timeoutMs, () => {
                req.destroy(new Error(`RemoteMailmanClient: request timed out after ${this.timeoutMs}ms`));
            });
            if (body)
                req.write(body);
            req.end();
        });
    }
}
exports.RemoteMailmanClient = RemoteMailmanClient;
