// @marsulta/mailman/dist/transport/server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MailmanServer = void 0;
const http_1 = __importDefault(require("http"));
// ─────────────────────────────────────────────
//  MailmanServer
//
//  Wraps an existing Runtime and exposes it over
//  HTTP so remote agents can send packets to it.
//
//  Endpoints:
//    POST /packets  — send a packet, get a reply
//    GET  /roles    — list registered roles
//    GET  /health   — runtime status + stats
//    GET  /metrics  — Prometheus-format counters
// ─────────────────────────────────────────────
class MailmanServer {
    runtime;
    server;
    port;
    host;
    authConfig;
    constructor(runtime, config = {}) {
        this.runtime = runtime;
        this.port = config.port ?? 7337;
        this.host = config.host ?? "127.0.0.1";
        this.authConfig = config.auth;
        this.server = http_1.default.createServer(this.handleRequest.bind(this));
    }
    // ── Lifecycle ─────────────────────────────
    listen() {
        return new Promise((resolve) => {
            this.server.listen(this.port, this.host, () => resolve());
        });
    }
    close() {
        return new Promise((resolve, reject) => {
            this.server.close((err) => (err ? reject(err) : resolve()));
        });
    }
    get address() {
        return `http://${this.host}:${this.port}`;
    }
    /** Expose the underlying http.Server (e.g. for WebSocket upgrade sharing). */
    get httpServer() {
        return this.server;
    }
    // ── Request handler ───────────────────────
    async handleRequest(req, res) {
        const url = req.url ?? "/";
        const method = req.method ?? "GET";
        // Auth check (skip for /health so monitors don't need keys)
        if (url !== "/health" && !this.isAuthorized(req)) {
            return sendJson(res, 401, { error: "Unauthorized — supply a valid API key" });
        }
        try {
            // POST /packets
            if (method === "POST" && url === "/packets") {
                const body = await readBody(req);
                const packet = JSON.parse(body);
                const reply = await this.runtime.send(packet);
                return sendJson(res, 200, reply);
            }
            // POST /dispatch — fire-and-forget
            if (method === "POST" && url === "/dispatch") {
                const body = await readBody(req);
                const packet = JSON.parse(body);
                this.runtime.dispatch(packet);
                return sendJson(res, 202, { accepted: true });
            }
            // GET /roles
            if (method === "GET" && url === "/roles") {
                return sendJson(res, 200, this.runtime.listRoles());
            }
            // GET /health
            if (method === "GET" && url === "/health") {
                return sendJson(res, 200, {
                    running: this.runtime.isRunning(),
                    stats: this.runtime.getStats(),
                    roles: this.runtime.listRoles().map((r) => r.name),
                });
            }
            // GET /metrics — Prometheus text format
            if (method === "GET" && url === "/metrics") {
                const stats = this.runtime.getStats();
                const lines = [
                    "# HELP mailman_packets_processed_total Total packets successfully processed",
                    "# TYPE mailman_packets_processed_total counter",
                    `mailman_packets_processed_total ${stats.packetsProcessed}`,
                    "# HELP mailman_packets_failed_total Total packets that failed (including dead-lettered)",
                    "# TYPE mailman_packets_failed_total counter",
                    `mailman_packets_failed_total ${stats.packetsFailed}`,
                    "# HELP mailman_packets_retried_total Total retry attempts across all packets",
                    "# TYPE mailman_packets_retried_total counter",
                    `mailman_packets_retried_total ${stats.packetsRetried}`,
                    "# HELP mailman_packets_dead_lettered_total Total packets sent to the dead-letter queue",
                    "# TYPE mailman_packets_dead_lettered_total counter",
                    `mailman_packets_dead_lettered_total ${stats.packetsDeadLettered}`,
                    "# HELP mailman_packets_dispatched_total Total fire-and-forget dispatches",
                    "# TYPE mailman_packets_dispatched_total counter",
                    `mailman_packets_dispatched_total ${stats.packetsDispatched}`,
                    "# HELP mailman_packets_published_total Total pub/sub publishes",
                    "# TYPE mailman_packets_published_total counter",
                    `mailman_packets_published_total ${stats.packetsPublished}`,
                    "# HELP mailman_roles_registered Current number of registered roles",
                    "# TYPE mailman_roles_registered gauge",
                    `mailman_roles_registered ${this.runtime.listRoles().length}`,
                    "# HELP mailman_dlq_size Current dead-letter queue depth",
                    "# TYPE mailman_dlq_size gauge",
                    `mailman_dlq_size ${this.runtime.getDLQ().count()}`,
                    "",
                ];
                res.writeHead(200, { "Content-Type": "text/plain; version=0.0.4" });
                res.end(lines.join("\n"));
                return;
            }
            // 404
            sendJson(res, 404, { error: `Unknown route: ${method} ${url}` });
        }
        catch (err) {
            const message = err instanceof Error ? err.message : "Internal server error";
            sendJson(res, 500, { error: message });
        }
    }
    // ── Auth ──────────────────────────────────
    isAuthorized(req) {
        if (!this.authConfig)
            return true;
        const headerName = (this.authConfig.header ?? "x-api-key").toLowerCase();
        const raw = req.headers[headerName];
        const provided = Array.isArray(raw) ? raw[0] : raw;
        if (provided === this.authConfig.apiKey)
            return true;
        // Also accept "Authorization: Bearer <key>"
        const authHeader = req.headers["authorization"];
        if (authHeader?.startsWith("Bearer ")) {
            return authHeader.slice(7) === this.authConfig.apiKey;
        }
        return false;
    }
}
exports.MailmanServer = MailmanServer;
// ─────────────────────────────────────────────
//  Helpers
// ─────────────────────────────────────────────
function readBody(req) {
    return new Promise((resolve, reject) => {
        const chunks = [];
        req.on("data", (chunk) => chunks.push(chunk));
        req.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
        req.on("error", reject);
    });
}
function sendJson(res, status, body) {
    const json = JSON.stringify(body);
    res.writeHead(status, {
        "Content-Type": "application/json",
        "Content-Length": Buffer.byteLength(json),
    });
    res.end(json);
}
