// @marsulta/mailman/dist/transport/wsServer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MailmanWsServer = void 0;
const ws_1 = require("ws");
// ─────────────────────────────────────────────
//  MailmanWsServer
// ─────────────────────────────────────────────
class MailmanWsServer {
    runtime;
    wss;
    port;
    host;
    apiKey;
    /** subscriptionId → unsub function (per-connection) */
    connSubs = new WeakMap();
    constructor(runtime, config = {}) {
        this.runtime = runtime;
        this.port = config.port ?? 7338;
        this.host = config.host ?? "127.0.0.1";
        this.apiKey = config.apiKey;
        if (config.server) {
            this.wss = new ws_1.WebSocketServer({ server: config.server });
        }
        else {
            this.wss = new ws_1.WebSocketServer({ port: this.port, host: this.host });
        }
        this.wss.on("connection", this.onConnection.bind(this));
    }
    // ── Lifecycle ─────────────────────────────
    /** Only needed when not attached to an existing http.Server. */
    close() {
        return new Promise((resolve, reject) => {
            this.wss.close((err) => (err ? reject(err) : resolve()));
        });
    }
    get clientCount() {
        return this.wss.clients.size;
    }
    // ── Connection handler ────────────────────
    onConnection(ws, req) {
        // Auth check
        if (this.apiKey && !this.checkAuth(req)) {
            ws.close(4001, "Unauthorized");
            return;
        }
        this.connSubs.set(ws, new Map());
        ws.on("message", (raw) => {
            this.handleMessage(ws, raw.toString()).catch((err) => {
                this.sendError(ws, undefined, "INTERNAL", err instanceof Error ? err.message : String(err));
            });
        });
        ws.on("close", () => {
            // Clean up all subscriptions for this connection
            const subs = this.connSubs.get(ws);
            if (subs) {
                for (const unsub of subs.values())
                    unsub();
                this.connSubs.delete(ws);
            }
        });
    }
    async handleMessage(ws, raw) {
        let msg;
        try {
            msg = JSON.parse(raw);
        }
        catch {
            this.sendError(ws, undefined, "PARSE_ERROR", "Invalid JSON");
            return;
        }
        const { op, requestId } = msg;
        switch (op) {
            case "send": {
                try {
                    const reply = await this.runtime.send(msg.packet);
                    this.send(ws, { op: "reply", requestId, packet: reply });
                }
                catch (err) {
                    this.sendError(ws, requestId, "SEND_FAILED", err instanceof Error ? err.message : String(err));
                }
                break;
            }
            case "dispatch": {
                this.runtime.dispatch(msg.packet);
                this.send(ws, { op: "ack", requestId });
                break;
            }
            case "subscribe": {
                const subs = this.connSubs.get(ws);
                if (subs.has(msg.topic)) {
                    // Already subscribed — no-op
                    this.send(ws, { op: "ack", requestId });
                    break;
                }
                const unsub = this.runtime.subscribe(msg.topic, (packet) => {
                    if (ws.readyState === ws_1.WebSocket.OPEN) {
                        this.send(ws, { op: "event", packet });
                    }
                });
                subs.set(msg.topic, unsub);
                this.send(ws, { op: "ack", requestId });
                break;
            }
            case "unsubscribe": {
                const subs = this.connSubs.get(ws);
                const unsub = subs.get(msg.topic);
                if (unsub) {
                    unsub();
                    subs.delete(msg.topic);
                }
                this.send(ws, { op: "ack", requestId });
                break;
            }
            default: {
                this.sendError(ws, msg.requestId, "UNKNOWN_OP", `Unknown op: ${op}`);
            }
        }
    }
    // ── Helpers ───────────────────────────────
    send(ws, payload) {
        if (ws.readyState === ws_1.WebSocket.OPEN) {
            ws.send(JSON.stringify(payload));
        }
    }
    sendError(ws, requestId, code, message) {
        this.send(ws, { op: "error", requestId, code, message });
    }
    checkAuth(req) {
        if (!this.apiKey)
            return true;
        // Header: x-api-key
        const header = req.headers["x-api-key"];
        if (header === this.apiKey)
            return true;
        // Bearer token
        const auth = req.headers["authorization"];
        if (auth?.startsWith("Bearer ") && auth.slice(7) === this.apiKey)
            return true;
        // Query param: ?apiKey=...
        const url = new URL(req.url ?? "/", "http://localhost");
        if (url.searchParams.get("apiKey") === this.apiKey)
            return true;
        return false;
    }
}
exports.MailmanWsServer = MailmanWsServer;
