// @marsulta/mailman/dist/experimental/ahp-test-harness/__tests__/localTransportFaults.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const WorkbenchTestHarness_1 = require("../WorkbenchTestHarness");
const scenarioSummaries = [];
const liveHarnesses = [];
function recordScenario(summary) {
    scenarioSummaries.push(summary);
}
async function waitForEvent(events, predicate, timeoutMs = 1000) {
    const started = Date.now();
    while (Date.now() - started < timeoutMs) {
        const found = events.find(predicate);
        if (found) {
            return found;
        }
        await new Promise((resolve) => setTimeout(resolve, 5));
    }
    throw new Error("Timed out waiting for fault-injection event");
}
function assertTaskConsistency(taskId, events) {
    (0, vitest_1.expect)(events.length).toBeGreaterThan(0);
    (0, vitest_1.expect)(events.every((event) => event.taskId === taskId)).toBe(true);
    (0, vitest_1.expect)(events.every((event) => event.correlationId === taskId)).toBe(true);
    (0, vitest_1.expect)(events.map((event) => event.sequence)).toEqual(events.map((_event, index) => index + 1));
}
async function makeFaultHarness(params) {
    const harness = (0, WorkbenchTestHarness_1.createAHPTestHarness)({
        enabled: true,
        transport: "local",
        timeouts: {
            handshakeTimeoutMs: 120,
            taskAckTimeoutMs: 40,
            inactivityTimeoutMs: 250,
            maxIdempotentRetries: 2,
            ...(params.timeouts ?? {}),
        },
        faultInjection: {
            seed: 1337,
            rules: params.rules,
        },
    });
    if (!harness) {
        throw new Error("Expected local AHP test harness");
    }
    liveHarnesses.push(harness);
    return harness;
}
(0, vitest_1.afterEach)(async () => {
    await Promise.all(liveHarnesses.splice(0).map((harness) => harness.close()));
});
(0, vitest_1.afterAll)(() => {
    console.table(scenarioSummaries);
});
(0, vitest_1.describe)("experimental AHP fault injection over local transport", () => {
    (0, vitest_1.it)("handles delayed delivery deterministically without breaking correlation or sequence guarantees", async () => {
        const taskId = "fault-delay-1";
        const harness = await makeFaultHarness({
            rules: [
                { id: "delay-partial-1", direction: "ahp->workbench", kind: "task.partial_output", taskId, occurrence: 1, delayMs: 20 },
                { id: "delay-partial-2", direction: "ahp->workbench", kind: "task.partial_output", taskId, occurrence: 2, delayMs: 40 },
                { id: "delay-complete", direction: "ahp->workbench", kind: "task.completed", taskId, delayMs: 60 },
            ],
        });
        const events = [];
        harness.maestro.onEvent((event) => events.push(event));
        await harness.maestro.connect({ sessionId: "fault-delay-session" });
        await harness.maestro.registerAgent({
            agentId: "fault-delay-agent",
            handler: (_envelope, context) => {
                context.accept({ stage: "accepted" });
                context.partialOutput({ chunk: "A" });
                context.partialOutput({ chunk: "B" });
                context.complete({ output: "AB" });
            },
        });
        await harness.maestro.sendTask({
            taskId,
            agentId: "fault-delay-agent",
            prompt: "delay test",
        });
        await waitForEvent(events, (event) => event.taskId === taskId && event.eventType === "TASK_COMPLETED");
        const taskEvents = events.filter((event) => event.taskId === taskId);
        const trace = harness.traceTask(taskId);
        (0, vitest_1.expect)(taskEvents.map((event) => event.eventType)).toEqual([
            "TASK_ACCEPTED",
            "TASK_PARTIAL_OUTPUT",
            "TASK_PARTIAL_OUTPUT",
            "TASK_COMPLETED",
        ]);
        (0, vitest_1.expect)(taskEvents.map((event) => event.payload?.chunk ?? event.payload?.output ?? event.payload?.stage)).toEqual([
            "accepted",
            "A",
            "B",
            "AB",
        ]);
        assertTaskConsistency(taskId, taskEvents);
        (0, vitest_1.expect)(trace.some((entry) => entry.eventType.startsWith("fault.delay-partial-1"))).toBe(true);
        (0, vitest_1.expect)(trace.some((entry) => entry.eventType.startsWith("fault.delay-complete"))).toBe(true);
        recordScenario({
            scenario: "delayed delivery",
            expected: "accepted -> partial -> partial -> completed, consistent correlation and contiguous sequence",
            observed: taskEvents.map((event) => event.eventType).join(" -> "),
            pass: "PASS",
        });
    });
    (0, vitest_1.it)("keeps cancel and duplicate handling idempotent under duplicated envelopes", async () => {
        const taskId = "fault-duplicate-cancel-1";
        const harness = await makeFaultHarness({
            rules: [
                { id: "dup-accept", direction: "ahp->workbench", kind: "task.accept", taskId, duplicateCount: 1, duplicateSpacingMs: 5 },
                { id: "dup-cancelled", direction: "ahp->workbench", kind: "task.cancelled", taskId, duplicateCount: 1, duplicateSpacingMs: 5 },
            ],
        });
        const events = [];
        harness.maestro.onEvent((event) => events.push(event));
        await harness.maestro.connect({ sessionId: "fault-duplicate-session" });
        await harness.maestro.registerAgent({
            agentId: "fault-duplicate-agent",
            handler: async (_envelope, context) => {
                context.accept({ stage: "accepted" });
                await new Promise((resolve) => {
                    context.signal.addEventListener("abort", () => {
                        context.cancelled({ reason: "cancelled" });
                        resolve();
                    }, { once: true });
                });
            },
        });
        await harness.maestro.sendTask({
            taskId,
            agentId: "fault-duplicate-agent",
            prompt: "duplicate cancel",
        });
        await waitForEvent(events, (event) => event.taskId === taskId && event.eventType === "TASK_ACCEPTED");
        await harness.maestro.cancelTask(taskId);
        await harness.maestro.cancelTask(taskId);
        await waitForEvent(events, (event) => event.taskId === taskId && event.eventType === "TASK_CANCELLED");
        await new Promise((resolve) => setTimeout(resolve, 30));
        const taskEvents = events.filter((event) => event.taskId === taskId);
        const trace = harness.traceTask(taskId);
        (0, vitest_1.expect)(taskEvents.filter((event) => event.eventType === "TASK_ACCEPTED")).toHaveLength(1);
        (0, vitest_1.expect)(taskEvents.filter((event) => event.eventType === "TASK_CANCELLED")).toHaveLength(1);
        assertTaskConsistency(taskId, taskEvents);
        (0, vitest_1.expect)(trace.some((entry) => entry.eventType.startsWith("fault.dup-accept"))).toBe(true);
        (0, vitest_1.expect)(trace.some((entry) => entry.eventType.startsWith("fault.dup-cancelled"))).toBe(true);
        recordScenario({
            scenario: "duplicated envelopes + idempotent cancel",
            expected: "single TASK_ACCEPTED and single TASK_CANCELLED despite duplicate frames and double cancel",
            observed: taskEvents.map((event) => event.eventType).join(" -> "),
            pass: "PASS",
        });
    });
    (0, vitest_1.it)("survives dropped non-terminal envelopes and still completes with consistent IDs", async () => {
        const taskId = "fault-drop-1";
        const harness = await makeFaultHarness({
            rules: [
                { id: "drop-progress", direction: "ahp->workbench", kind: "task.progress", taskId, drop: true },
            ],
        });
        const events = [];
        harness.maestro.onEvent((event) => events.push(event));
        await harness.maestro.connect({ sessionId: "fault-drop-session" });
        await harness.maestro.registerAgent({
            agentId: "fault-drop-agent",
            handler: (_envelope, context) => {
                context.accept({ stage: "accepted" });
                context.progress({ percent: 10 });
                context.partialOutput({ chunk: "chunk" });
                context.complete({ output: "done" });
            },
        });
        await harness.maestro.sendTask({
            taskId,
            agentId: "fault-drop-agent",
            prompt: "drop progress",
        });
        await waitForEvent(events, (event) => event.taskId === taskId && event.eventType === "TASK_COMPLETED");
        const taskEvents = events.filter((event) => event.taskId === taskId);
        const trace = harness.traceTask(taskId);
        (0, vitest_1.expect)(taskEvents.map((event) => event.eventType)).toEqual([
            "TASK_ACCEPTED",
            "TASK_PARTIAL_OUTPUT",
            "TASK_COMPLETED",
        ]);
        assertTaskConsistency(taskId, taskEvents);
        (0, vitest_1.expect)(trace.some((entry) => entry.eventType.startsWith("fault.drop-progress"))).toBe(true);
        recordScenario({
            scenario: "dropped non-terminal envelope",
            expected: "progress missing but task still completes once with consistent IDs",
            observed: taskEvents.map((event) => event.eventType).join(" -> "),
            pass: "PASS",
        });
    });
    (0, vitest_1.it)("enforces terminal-state monotonicity under out-of-order delivery", async () => {
        const taskId = "fault-out-of-order-1";
        const harness = await makeFaultHarness({
            rules: [
                { id: "delay-accept", direction: "ahp->workbench", kind: "task.accept", taskId, delayMs: 40 },
                { id: "delay-progress", direction: "ahp->workbench", kind: "task.progress", taskId, delayMs: 55 },
            ],
        });
        const events = [];
        harness.maestro.onEvent((event) => events.push(event));
        await harness.maestro.connect({ sessionId: "fault-out-of-order-session" });
        await harness.maestro.registerAgent({
            agentId: "fault-out-of-order-agent",
            handler: (_envelope, context) => {
                context.accept({ stage: "accepted" });
                context.progress({ percent: 50 });
                context.complete({ output: "done-first" });
            },
        });
        await harness.maestro.sendTask({
            taskId,
            agentId: "fault-out-of-order-agent",
            prompt: "out of order",
        });
        await waitForEvent(events, (event) => event.taskId === taskId && event.eventType === "TASK_COMPLETED");
        await new Promise((resolve) => setTimeout(resolve, 80));
        const taskEvents = events.filter((event) => event.taskId === taskId);
        const trace = harness.traceTask(taskId);
        (0, vitest_1.expect)(taskEvents.map((event) => event.eventType)).toEqual(["TASK_COMPLETED"]);
        assertTaskConsistency(taskId, taskEvents);
        (0, vitest_1.expect)(trace.some((entry) => entry.eventType.startsWith("fault.delay-accept"))).toBe(true);
        (0, vitest_1.expect)(trace.some((entry) => entry.eventType.startsWith("fault.delay-progress"))).toBe(true);
        (0, vitest_1.expect)(trace.filter((entry) => entry.direction === "ahp->workbench" && entry.eventType === "task.accept").length).toBeGreaterThan(0);
        recordScenario({
            scenario: "out-of-order delivery",
            expected: "completed observed once; late non-terminal events ignored after terminal state",
            observed: taskEvents.map((event) => event.eventType).join(" -> "),
            pass: "PASS",
        });
    });
    (0, vitest_1.it)("recovers from a transient disconnect and reconnect on the real transport", async () => {
        const taskId = "fault-disconnect-1";
        const secondTaskId = "fault-disconnect-2";
        const harness = await makeFaultHarness({
            rules: [
                {
                    id: "disconnect-on-accept",
                    direction: "ahp->workbench",
                    kind: "task.accept",
                    taskId,
                    disconnectBeforeDeliver: true,
                    reconnectAfterMs: 25,
                    delayMs: 5,
                },
                {
                    id: "delay-complete-after-reconnect",
                    direction: "ahp->workbench",
                    kind: "task.completed",
                    taskId,
                    delayMs: 45,
                },
            ],
            timeouts: {
                taskAckTimeoutMs: 90,
                inactivityTimeoutMs: 300,
            },
        });
        const events = [];
        harness.maestro.onEvent((event) => events.push(event));
        await harness.maestro.connect({ sessionId: "fault-disconnect-session" });
        await harness.maestro.registerAgent({
            agentId: "fault-disconnect-agent",
            handler: (_envelope, context) => {
                context.accept({ stage: "accepted" });
                context.complete({ output: "ok" });
            },
        });
        await harness.maestro.sendTask({
            taskId,
            agentId: "fault-disconnect-agent",
            prompt: "disconnect once",
        });
        await waitForEvent(events, (event) => event.taskId === taskId && event.eventType === "TASK_COMPLETED");
        await harness.maestro.sendTask({
            taskId: secondTaskId,
            agentId: "fault-disconnect-agent",
            prompt: "post reconnect",
        });
        await waitForEvent(events, (event) => event.taskId === secondTaskId && event.eventType === "TASK_COMPLETED");
        const firstTaskEvents = events.filter((event) => event.taskId === taskId);
        const secondTaskEvents = events.filter((event) => event.taskId === secondTaskId);
        const trace = harness.traceTask(taskId);
        (0, vitest_1.expect)(firstTaskEvents.map((event) => event.eventType)).toEqual([
            "TASK_ACCEPTED",
            "TASK_COMPLETED",
        ]);
        (0, vitest_1.expect)(secondTaskEvents.map((event) => event.eventType)).toEqual([
            "TASK_ACCEPTED",
            "TASK_COMPLETED",
        ]);
        assertTaskConsistency(taskId, firstTaskEvents);
        assertTaskConsistency(secondTaskId, secondTaskEvents);
        (0, vitest_1.expect)(trace.some((entry) => entry.eventType.startsWith("fault.disconnect-on-accept"))).toBe(true);
        recordScenario({
            scenario: "transient disconnect/reconnect",
            expected: "first task survives socket flap; second task completes after reconnect",
            observed: `${firstTaskEvents.map((event) => event.eventType).join(" -> ")} | ${secondTaskEvents.map((event) => event.eventType).join(" -> ")}`,
            pass: "PASS",
        });
    });
    (0, vitest_1.it)("handles retry race windows without producing duplicate completions", async () => {
        const taskId = "fault-retry-race-1";
        const harness = await makeFaultHarness({
            rules: [
                // Accept is delayed beyond one ack timeout (50 ms) so exactly one retry fires,
                // but arrives well before the second ack timeout (~102 ms) giving a ~30 ms margin.
                { id: "delay-accept-for-retry", direction: "ahp->workbench", kind: "task.accept", taskId, delayMs: 70 },
                // Complete follows accept so the sequence order is preserved.
                { id: "delay-complete-for-retry", direction: "ahp->workbench", kind: "task.completed", taskId, delayMs: 85 },
            ],
            timeouts: {
                taskAckTimeoutMs: 50,
                inactivityTimeoutMs: 600,
                maxIdempotentRetries: 2,
            },
        });
        const events = [];
        harness.maestro.onEvent((event) => events.push(event));
        await harness.maestro.connect({ sessionId: "fault-retry-session" });
        await harness.maestro.registerAgent({
            agentId: "fault-retry-agent",
            handler: (_envelope, context) => {
                context.accept({ stage: "accepted" });
                context.complete({ output: "once" });
            },
        });
        await harness.maestro.sendTask({
            taskId,
            agentId: "fault-retry-agent",
            prompt: "retry race",
        });
        await waitForEvent(events, (event) => event.taskId === taskId && event.eventType === "TASK_COMPLETED", 1500);
        await new Promise((resolve) => setTimeout(resolve, 120));
        const taskEvents = events.filter((event) => event.taskId === taskId);
        const trace = harness.traceTask(taskId);
        const sendCount = trace.filter((entry) => entry.direction === "workbench->ahp" && entry.eventType === "task.send").length;
        (0, vitest_1.expect)(taskEvents.filter((event) => event.eventType === "TASK_ACCEPTED")).toHaveLength(1);
        (0, vitest_1.expect)(taskEvents.filter((event) => event.eventType === "TASK_COMPLETED")).toHaveLength(1);
        (0, vitest_1.expect)(taskEvents.filter((event) => event.eventType === "TASK_FAILED")).toHaveLength(0);
        (0, vitest_1.expect)(sendCount).toBe(2);
        assertTaskConsistency(taskId, taskEvents);
        recordScenario({
            scenario: "retry race window",
            expected: "one retry send, one completion, no duplicate terminal event",
            observed: `${taskEvents.map((event) => event.eventType).join(" -> ")} (sends=${sendCount})`,
            pass: "PASS",
        });
    });
});
