// @marsulta/mailman/dist/index.js
"use strict";
// ─────────────────────────────────────────────
//  Mailman — public API
// ─────────────────────────────────────────────
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.elapsedMs = exports.now = exports.newId = exports.newTaskId = exports.newPacketId = exports.createClient = exports.MailmanClient = exports.MailmanWsServer = exports.RemoteMailmanClient = exports.MailmanServer = exports.PacketScheduler = exports.LoadBalancer = exports.StreamSessionStore = exports.StreamSession = exports.EventBus = exports.MemoryDLQStore = exports.MemoryTraceStore = exports.DEFAULT_DB_PATH = exports.DEFAULT_RETRY_POLICY = exports.ErrorCode = exports.MailmanError = exports.DeadLetterQueue = exports.TraceStore = exports.Validator = exports.Router = exports.Registry = exports.resetRuntime = exports.getRuntime = exports.Runtime = void 0;
// Packet types + factories
__exportStar(require("./packet/types"), exports);
__exportStar(require("./packet/createPacket"), exports);
// Core
var runtime_1 = require("./core/runtime");
Object.defineProperty(exports, "Runtime", { enumerable: true, get: function () { return runtime_1.Runtime; } });
Object.defineProperty(exports, "getRuntime", { enumerable: true, get: function () { return runtime_1.getRuntime; } });
Object.defineProperty(exports, "resetRuntime", { enumerable: true, get: function () { return runtime_1.resetRuntime; } });
var registry_1 = require("./core/registry");
Object.defineProperty(exports, "Registry", { enumerable: true, get: function () { return registry_1.Registry; } });
var router_1 = require("./core/router");
Object.defineProperty(exports, "Router", { enumerable: true, get: function () { return router_1.Router; } });
var validator_1 = require("./core/validator");
Object.defineProperty(exports, "Validator", { enumerable: true, get: function () { return validator_1.Validator; } });
var traceStore_1 = require("./core/traceStore");
Object.defineProperty(exports, "TraceStore", { enumerable: true, get: function () { return traceStore_1.TraceStore; } });
var dlq_1 = require("./core/dlq");
Object.defineProperty(exports, "DeadLetterQueue", { enumerable: true, get: function () { return dlq_1.DeadLetterQueue; } });
var errors_1 = require("./core/errors");
Object.defineProperty(exports, "MailmanError", { enumerable: true, get: function () { return errors_1.MailmanError; } });
Object.defineProperty(exports, "ErrorCode", { enumerable: true, get: function () { return errors_1.ErrorCode; } });
var retry_1 = require("./core/retry");
Object.defineProperty(exports, "DEFAULT_RETRY_POLICY", { enumerable: true, get: function () { return retry_1.DEFAULT_RETRY_POLICY; } });
var db_1 = require("./core/db");
Object.defineProperty(exports, "DEFAULT_DB_PATH", { enumerable: true, get: function () { return db_1.DEFAULT_DB_PATH; } });
// Pluggable backends (Item 10)
var backends_1 = require("./core/backends");
Object.defineProperty(exports, "MemoryTraceStore", { enumerable: true, get: function () { return backends_1.MemoryTraceStore; } });
Object.defineProperty(exports, "MemoryDLQStore", { enumerable: true, get: function () { return backends_1.MemoryDLQStore; } });
// Pub/Sub (Item 2)
var eventBus_1 = require("./core/eventBus");
Object.defineProperty(exports, "EventBus", { enumerable: true, get: function () { return eventBus_1.EventBus; } });
// Streaming (Item 4)
var streamSession_1 = require("./core/streamSession");
Object.defineProperty(exports, "StreamSession", { enumerable: true, get: function () { return streamSession_1.StreamSession; } });
Object.defineProperty(exports, "StreamSessionStore", { enumerable: true, get: function () { return streamSession_1.StreamSessionStore; } });
// Capability routing (Item 6)
var loadBalancer_1 = require("./core/loadBalancer");
Object.defineProperty(exports, "LoadBalancer", { enumerable: true, get: function () { return loadBalancer_1.LoadBalancer; } });
// Scheduling (Item 8)
var scheduler_1 = require("./core/scheduler");
Object.defineProperty(exports, "PacketScheduler", { enumerable: true, get: function () { return scheduler_1.PacketScheduler; } });
// Transport
var server_1 = require("./transport/server");
Object.defineProperty(exports, "MailmanServer", { enumerable: true, get: function () { return server_1.MailmanServer; } });
var remoteClient_1 = require("./transport/remoteClient");
Object.defineProperty(exports, "RemoteMailmanClient", { enumerable: true, get: function () { return remoteClient_1.RemoteMailmanClient; } });
var wsServer_1 = require("./transport/wsServer");
Object.defineProperty(exports, "MailmanWsServer", { enumerable: true, get: function () { return wsServer_1.MailmanWsServer; } });
// Client
var client_1 = require("./client/client");
Object.defineProperty(exports, "MailmanClient", { enumerable: true, get: function () { return client_1.MailmanClient; } });
Object.defineProperty(exports, "createClient", { enumerable: true, get: function () { return client_1.createClient; } });
// Utils
var ids_1 = require("./utils/ids");
Object.defineProperty(exports, "newPacketId", { enumerable: true, get: function () { return ids_1.newPacketId; } });
Object.defineProperty(exports, "newTaskId", { enumerable: true, get: function () { return ids_1.newTaskId; } });
Object.defineProperty(exports, "newId", { enumerable: true, get: function () { return ids_1.newId; } });
var time_1 = require("./utils/time");
Object.defineProperty(exports, "now", { enumerable: true, get: function () { return time_1.now; } });
Object.defineProperty(exports, "elapsedMs", { enumerable: true, get: function () { return time_1.elapsedMs; } });
