// @marsulta/mailman/dist/core/backends.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemoryDLQStore = exports.MemoryTraceStore = void 0;
const ids_1 = require("../utils/ids");
const time_1 = require("../utils/time");
// ─────────────────────────────────────────────
//  MemoryTraceStore
//
//  Zero-dependency in-memory implementation.
//  Ideal for tests or short-lived processes.
// ─────────────────────────────────────────────
class MemoryTraceStore {
    entries = [];
    record(packetId, taskId, event, actor, details) {
        this.entries.push({
            traceId: (0, ids_1.newId)("trace"),
            packetId,
            taskId,
            event,
            actor,
            timestamp: (0, time_1.now)(),
            details,
        });
    }
    getByPacket(packetId) {
        return this.entries.filter((e) => e.packetId === packetId);
    }
    getByTask(taskId) {
        return this.entries.filter((e) => e.taskId === taskId);
    }
    all() {
        return [...this.entries];
    }
    totalEvents() {
        return this.entries.length;
    }
    query({ packetId, taskId, event, actor, since, until, limit = 100 }) {
        let results = this.entries;
        if (packetId)
            results = results.filter((e) => e.packetId === packetId);
        if (taskId)
            results = results.filter((e) => e.taskId === taskId);
        if (event)
            results = results.filter((e) => e.event === event);
        if (actor)
            results = results.filter((e) => e.actor === actor);
        if (since)
            results = results.filter((e) => e.timestamp >= since);
        if (until)
            results = results.filter((e) => e.timestamp <= until);
        return results.slice(-limit);
    }
    clear() {
        this.entries.length = 0;
    }
}
exports.MemoryTraceStore = MemoryTraceStore;
// ─────────────────────────────────────────────
//  MemoryDLQStore
// ─────────────────────────────────────────────
class MemoryDLQStore {
    entries = new Map();
    push(packet, error, attempts) {
        // If the packet is already in the DLQ update it in-place
        for (const entry of this.entries.values()) {
            if (entry.packet.packetId === packet.packetId) {
                entry.error = error;
                entry.attempts = attempts;
                entry.lastSeen = (0, time_1.now)();
                return;
            }
        }
        const id = (0, ids_1.newId)("dlq");
        const ts = (0, time_1.now)();
        this.entries.set(id, { id, packet, error, attempts, firstSeen: ts, lastSeen: ts });
    }
    list() {
        return Array.from(this.entries.values());
    }
    get(dlqId) {
        return this.entries.get(dlqId);
    }
    count() {
        return this.entries.size;
    }
    remove(dlqId) {
        return this.entries.delete(dlqId);
    }
    clear() {
        this.entries.clear();
    }
}
exports.MemoryDLQStore = MemoryDLQStore;
