// @marsulta/mailman/dist/core/dlq.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeadLetterQueue = void 0;
const ids_1 = require("../utils/ids");
const time_1 = require("../utils/time");
const db_1 = require("./db");
// ─────────────────────────────────────────────
//  DeadLetterQueue — SQLite-backed
//
//  Lives in the same DB as TraceStore so
//  Postmaster can see both in one file.
// ─────────────────────────────────────────────
class DeadLetterQueue {
    db;
    /**
     * @param dbPath SQLite file path. Defaults to ~/.mailman/traces.db.
     *               Pass ':memory:' for tests.
     */
    constructor(dbPath) {
        this.db = (0, db_1.openDb)(dbPath);
    }
    // ── Write ──────────────────────────────────
    push(packet, error, attempts) {
        const id = (0, ids_1.newId)("dlq");
        const ts = (0, time_1.now)();
        this.db
            .prepare(`INSERT INTO dead_letter_queue
         (id, packet, error, attempts, first_seen, last_seen)
         VALUES (?, ?, ?, ?, ?, ?)`)
            .run(id, JSON.stringify(packet), error, attempts, ts, ts);
        return { id, packet, error, attempts, firstSeen: ts, lastSeen: ts };
    }
    // ── Read ───────────────────────────────────
    list() {
        return this.db
            .prepare("SELECT * FROM dead_letter_queue ORDER BY last_seen DESC")
            .all().map(rowToEntry);
    }
    get(id) {
        const row = this.db
            .prepare("SELECT * FROM dead_letter_queue WHERE id = ?")
            .get(id);
        return row ? rowToEntry(row) : undefined;
    }
    count() {
        const row = this.db
            .prepare("SELECT COUNT(*) as n FROM dead_letter_queue")
            .get();
        return row.n;
    }
    // ── Remove ─────────────────────────────────
    remove(id) {
        const result = this.db
            .prepare("DELETE FROM dead_letter_queue WHERE id = ?")
            .run(id);
        return result.changes > 0;
    }
    clear() {
        this.db.exec("DELETE FROM dead_letter_queue");
    }
}
exports.DeadLetterQueue = DeadLetterQueue;
// ─────────────────────────────────────────────
//  Row → domain type
// ─────────────────────────────────────────────
function rowToEntry(row) {
    return {
        id: row["id"],
        packet: JSON.parse(row["packet"]),
        error: row["error"],
        attempts: row["attempts"],
        firstSeen: row["first_seen"],
        lastSeen: row["last_seen"],
    };
}
