// @marsulta/mailman/dist/core/eventBus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventBus = void 0;
class EventBus {
    subs = new Map();
    /**
     * Subscribe to a topic pattern.
     * Returns an unsubscribe function.
     *
     * @example
     * const unsub = bus.subscribe("task.*", (pkt) => console.log(pkt));
     * // later:
     * unsub();
     */
    subscribe(topic, handler) {
        if (!this.subs.has(topic)) {
            this.subs.set(topic, new Set());
        }
        this.subs.get(topic).add(handler);
        return () => this.unsubscribe(topic, handler);
    }
    unsubscribe(topic, handler) {
        const set = this.subs.get(topic);
        if (!set)
            return;
        set.delete(handler);
        if (set.size === 0)
            this.subs.delete(topic);
    }
    /**
     * Deliver a packet to all matching subscribers.
     * Exceptions thrown by subscribers are swallowed — the bus never crashes.
     */
    publish(packet) {
        for (const [topic, handlers] of this.subs) {
            if (this.matches(topic, packet.type)) {
                for (const handler of handlers) {
                    try {
                        handler(packet);
                    }
                    catch {
                        // intentional: pub/sub is fire-and-forget
                    }
                }
            }
        }
    }
    /** Total number of active subscriptions across all topics. */
    subscriberCount(topic) {
        if (topic !== undefined)
            return this.subs.get(topic)?.size ?? 0;
        let total = 0;
        for (const set of this.subs.values())
            total += set.size;
        return total;
    }
    // ── Topic matching ────────────────────────
    matches(topic, packetType) {
        if (topic === "*")
            return true;
        if (topic === packetType)
            return true;
        // "task.*"  → matches "task.assign", "task.result", etc.
        // "stream.*" → matches "stream.open", "stream.chunk", etc.
        if (topic.endsWith(".*")) {
            const prefix = topic.slice(0, -2);
            return packetType === prefix || packetType.startsWith(prefix + ".");
        }
        return false;
    }
}
exports.EventBus = EventBus;
