// @marsulta/mailman/dist/core/registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Registry = void 0;
const errors_1 = require("./errors");
// ─────────────────────────────────────────────
//  Registry
// ─────────────────────────────────────────────
class Registry {
    entries = new Map();
    register(role, handler) {
        if (this.entries.has(role.name)) {
            throw new errors_1.MailmanError(errors_1.ErrorCode.ROLE_ALREADY_REGISTERED, `Role already registered: ${role.name}`);
        }
        this.entries.set(role.name, {
            registration: role,
            handler,
            registeredAt: new Date().toISOString(),
        });
    }
    unregister(name) {
        if (!this.entries.has(name)) {
            throw new errors_1.MailmanError(errors_1.ErrorCode.ROLE_NOT_FOUND, `Role not found: ${name}`);
        }
        this.entries.delete(name);
    }
    get(name) {
        return this.entries.get(name);
    }
    has(name) {
        return this.entries.has(name);
    }
    list() {
        return Array.from(this.entries.values()).map((e) => e.registration);
    }
    size() {
        return this.entries.size;
    }
    // ── Capability queries ────────────────────
    /**
     * Find all roles that advertise a specific capability.
     * Used by runtime.sendToCapability() for dynamic routing.
     */
    findByCapability(capability) {
        return Array.from(this.entries.values()).filter((e) => e.registration.capabilities?.includes(capability) ?? false);
    }
    /**
     * Find all roles that advertise ALL of the given capabilities.
     */
    findByCapabilities(capabilities) {
        return Array.from(this.entries.values()).filter((e) => capabilities.every((cap) => e.registration.capabilities?.includes(cap) ?? false));
    }
}
exports.Registry = Registry;
