// @marsulta/mailman/dist/core/retry.js
"use strict";
// ─────────────────────────────────────────────
//  Retry policy + exponential backoff helper
// ─────────────────────────────────────────────
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_RETRY_POLICY = void 0;
exports.withRetry = withRetry;
exports.DEFAULT_RETRY_POLICY = {
    maxAttempts: 3,
    baseDelayMs: 250,
    maxDelayMs: 5000,
    backoff: "exponential",
};
// ─────────────────────────────────────────────
//  Compute delay for attempt N (0-indexed)
// ─────────────────────────────────────────────
function computeDelay(attempt, policy) {
    if (policy.backoff === "fixed")
        return policy.baseDelayMs;
    // Exponential with ±30% jitter to spread thundering herds
    const exp = Math.min(policy.baseDelayMs * Math.pow(2, attempt), policy.maxDelayMs);
    const jitter = Math.random() * 0.3 * exp;
    return Math.floor(exp + jitter);
}
async function withRetry(fn, policy = exports.DEFAULT_RETRY_POLICY) {
    let lastError = new Error("Unknown error");
    for (let attempt = 0; attempt < policy.maxAttempts; attempt++) {
        try {
            const result = await fn();
            return { ok: true, result, attempts: attempt + 1 };
        }
        catch (err) {
            lastError = err instanceof Error ? err : new Error(String(err));
            // Don't sleep after the last attempt
            if (attempt < policy.maxAttempts - 1) {
                const delay = computeDelay(attempt, policy);
                await new Promise((resolve) => setTimeout(resolve, delay));
            }
        }
    }
    return { ok: false, error: lastError, attempts: policy.maxAttempts };
}
