// @marsulta/mailman/dist/core/runtime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Runtime = void 0;
exports.getRuntime = getRuntime;
exports.resetRuntime = resetRuntime;
const registry_1 = require("./registry");
const router_1 = require("./router");
const traceStore_1 = require("./traceStore");
const validator_1 = require("./validator");
const errors_1 = require("./errors");
const middleware_1 = require("./middleware");
const dlq_1 = require("./dlq");
const retry_1 = require("./retry");
const db_1 = require("./db");
const eventBus_1 = require("./eventBus");
const streamSession_1 = require("./streamSession");
const loadBalancer_1 = require("./loadBalancer");
const scheduler_1 = require("./scheduler");
const ids_1 = require("../utils/ids");
const time_1 = require("../utils/time");
// ─────────────────────────────────────────────
//  Runtime
// ─────────────────────────────────────────────
class Runtime {
    registry;
    router;
    traceStore;
    dlq;
    validator;
    middleware = [];
    retryPolicy;
    customTypes;
    // Item 2 — Pub/Sub
    eventBus;
    // Item 4 — Streaming
    streamSessions;
    // Item 6 — Load balancing
    loadBalancer;
    // Item 8 — Scheduling
    scheduler;
    // Item 7 — Ack/nack
    pendingAcks = new Map();
    // Item 9 — Telemetry
    telemetryHook;
    running = false;
    stats = {
        packetsProcessed: 0,
        packetsFailed: 0,
        packetsRetried: 0,
        packetsDeadLettered: 0,
        packetsDispatched: 0,
        packetsPublished: 0,
        startedAt: null,
    };
    constructor(config = {}) {
        const dbPath = config.dbPath ?? db_1.DEFAULT_DB_PATH;
        this.registry = new registry_1.Registry();
        this.router = new router_1.Router(this.registry);
        this.traceStore = config.traceStore ?? new traceStore_1.TraceStore(dbPath);
        this.dlq = config.dlq ?? new dlq_1.DeadLetterQueue(dbPath);
        this.customTypes = new Set(config.customTypes ?? []);
        this.validator = new validator_1.Validator(this.registry, this.customTypes);
        this.retryPolicy = { ...retry_1.DEFAULT_RETRY_POLICY, ...(config.retry ?? {}) };
        this.telemetryHook = config.onTelemetry;
        this.eventBus = new eventBus_1.EventBus();
        this.streamSessions = new streamSession_1.StreamSessionStore();
        this.loadBalancer = new loadBalancer_1.LoadBalancer();
        this.scheduler = new scheduler_1.PacketScheduler((pkt) => this.dispatch(pkt));
    }
    // ── Lifecycle ─────────────────────────────
    start() {
        if (this.running)
            return;
        this.running = true;
        this.stats.startedAt = (0, time_1.now)();
    }
    stop() {
        this.running = false;
    }
    isRunning() {
        return this.running;
    }
    // ── Middleware ────────────────────────────
    use(fn) {
        this.middleware.push(fn);
        return this;
    }
    // ── Role management ───────────────────────
    registerRole(role, handler) {
        this.registry.register(role, handler);
    }
    unregisterRole(name) {
        this.registry.unregister(name);
    }
    listRoles() {
        return this.registry.list();
    }
    // ── Item 1: Custom packet type registration ──
    /**
     * Register a user-defined packet type so the validator accepts it.
     * Use dot-notation namespacing: "myapp.event", "agent.thought", etc.
     */
    registerPacketType(type) {
        this.customTypes.add(type);
        return this;
    }
    /** Returns all currently known types (built-in + custom). */
    registeredTypes() {
        const { BUILTIN_PACKET_TYPES } = require("../packet/types");
        return [...BUILTIN_PACKET_TYPES, ...this.customTypes];
    }
    // ── Core: send ────────────────────────────
    async send(packet) {
        const { packetId, taskId, sender } = packet;
        this.record(packetId, taskId, "packet.received", sender);
        const result = this.validator.validate(packet);
        if (!result.ok) {
            this.record(packetId, taskId, "packet.rejected", "mailman", {
                reason: result.message,
                code: result.code,
            });
            this.stats.packetsFailed++;
            return this.buildErrorPacket(packet, result.code, result.message);
        }
        this.record(packetId, taskId, "packet.validated", "mailman");
        this.record(packetId, taskId, "packet.routed", "mailman", { target: packet.target });
        this.record(packetId, taskId, "packet.delivered", "mailman", { target: packet.target });
        this.record(packetId, taskId, "handler.started", packet.target);
        const coreDispatch = (pkt) => this.router.dispatch(pkt);
        const pipeline = (0, middleware_1.composeMiddleware)(this.middleware, coreDispatch);
        const retryPolicy = { ...this.retryPolicy };
        const outcome = await (0, retry_1.withRetry)(() => {
            const timeoutMs = packet.meta?.timeoutMs;
            if (timeoutMs && timeoutMs > 0) {
                return this.withTimeout(pipeline(packet), timeoutMs, packet);
            }
            return pipeline(packet);
        }, retryPolicy);
        if (!outcome.ok) {
            const err = outcome.error;
            const msg = err.message ?? "Handler threw an unknown error";
            const code = err instanceof errors_1.MailmanError ? err.code : errors_1.ErrorCode.HANDLER_THREW;
            const attempts = outcome.attempts;
            if (attempts > 1)
                this.stats.packetsRetried += attempts - 1;
            this.record(packetId, taskId, "handler.threw", packet.target, { error: msg, code, attempts });
            this.dlq.push(packet, msg, attempts);
            this.stats.packetsDeadLettered++;
            this.record(packetId, taskId, "packet.failed", "mailman", { deadLettered: true, attempts });
            this.stats.packetsFailed++;
            return this.buildErrorPacket(packet, code, msg, { attempts });
        }
        const response = outcome.result;
        if (outcome.attempts > 1) {
            this.stats.packetsRetried += outcome.attempts - 1;
            this.record(packetId, taskId, "handler.finished", packet.target, { attempts: outcome.attempts });
        }
        else {
            this.record(packetId, taskId, "handler.finished", packet.target);
        }
        this.record(packetId, taskId, "packet.completed", "mailman");
        this.stats.packetsProcessed++;
        // Track ack/nack for sendWithAck
        if (response.type === "task.ack") {
            this.record(packetId, taskId, "packet.acked", "mailman");
        }
        else if (response.type === "task.nack") {
            this.record(packetId, taskId, "packet.nacked", "mailman");
            // Reject any waiting ack result promise
            const pending = this.pendingAcks.get(taskId);
            if (pending) {
                pending.reject(new errors_1.MailmanError(errors_1.ErrorCode.HANDLER_THREW, response.error?.message ?? "Handler nacked"));
                this.pendingAcks.delete(taskId);
            }
        }
        // Fan out to pub/sub subscribers
        this.eventBus.publish(response);
        return response;
    }
    // ── Item 2: Fire-and-forget dispatch ──────
    /**
     * Send a packet without waiting for a reply.
     * Validation and routing still run; errors are traced + dead-lettered.
     */
    dispatch(packet) {
        this.record(packet.packetId, packet.taskId, "packet.dispatched", packet.sender);
        this.stats.packetsDispatched++;
        this.send(packet).catch(() => {
            // errors already traced and dead-lettered inside send()
        });
    }
    // ── Item 2: Pub/Sub ───────────────────────
    /**
     * Subscribe to packets by type pattern.
     * Topics: exact type, "task.*", or "*" for all.
     * Returns an unsubscribe function.
     */
    subscribe(topic, handler) {
        return this.eventBus.subscribe(topic, handler);
    }
    /**
     * Publish a packet directly to all matching subscribers.
     * No validation, no routing, no reply — pure event fan-out.
     */
    publish(packet) {
        this.record(packet.packetId, packet.taskId, "packet.published", packet.sender);
        this.stats.packetsPublished++;
        this.eventBus.publish(packet);
    }
    // ── Item 4: Streaming ─────────────────────
    /**
     * Register a role that produces a stream of chunks instead of a single reply.
     * The handler calls stream.push(chunk), stream.end(), or stream.error().
     *
     * @example
     * runtime.registerStreamRole(
     *   { name: "llm", accepts: ["stream.open"], capabilities: ["generate"] },
     *   async (packet, stream) => {
     *     for await (const token of llm.stream(packet.payload.prompt as string)) {
     *       stream.push(token);
     *     }
     *     stream.end();
     *   }
     * );
     */
    registerStreamRole(role, handler) {
        const wrappedHandler = async (packet) => {
            const streamId = packet.meta?.streamId;
            if (!streamId) {
                throw new errors_1.MailmanError(errors_1.ErrorCode.STREAM_NOT_FOUND, "No streamId in packet meta");
            }
            const session = this.streamSessions.get(streamId);
            if (!session) {
                throw new errors_1.MailmanError(errors_1.ErrorCode.STREAM_NOT_FOUND, `Stream session not found: ${streamId}`);
            }
            // Run stream handler asynchronously — don't block send()
            handler(packet, session).then(() => {
                if (!session.isEnded)
                    session.end();
                this.record(packet.packetId, packet.taskId, "stream.closed", role.name, { streamId });
                this.streamSessions.delete(streamId);
            }).catch((err) => {
                session.error(err);
                this.record(packet.packetId, packet.taskId, "stream.errored", role.name, {
                    streamId,
                    error: err.message,
                });
                this.streamSessions.delete(streamId);
            });
            // Return an immediate ack so send() can complete
            return {
                packetId: (0, ids_1.newPacketId)(),
                taskId: packet.taskId,
                parentPacketId: packet.packetId,
                type: "task.ack",
                sender: role.name,
                target: packet.sender,
                timestamp: (0, time_1.now)(),
                payload: { streaming: true, streamId },
            };
        };
        this.registry.register(role, wrappedHandler);
    }
    /**
     * Open a streaming session to a registered stream role.
     * Returns an AsyncIterable<string> that yields chunks as they arrive.
     *
     * @example
     * const chunks = await runtime.openStream({
     *   packetId: newPacketId(), taskId: newTaskId(),
     *   type: "stream.open", sender: "me", target: "llm",
     *   payload: { prompt: "Tell me a story" }, timestamp: now(),
     * });
     * for await (const chunk of chunks) process.stdout.write(chunk);
     */
    async openStream(packet) {
        const streamId = (0, ids_1.newId)("stream");
        const session = this.streamSessions.create(streamId);
        const augmented = {
            ...packet,
            meta: { ...packet.meta, streamId },
        };
        this.record(packet.packetId, packet.taskId, "stream.opened", packet.sender, { streamId });
        const ack = await this.send(augmented);
        if (ack.type === "error.report") {
            session.error(new Error(ack.error?.message ?? "Stream open failed"));
            this.streamSessions.delete(streamId);
        }
        return session;
    }
    // ── Item 6: Capability routing ────────────
    /**
     * Send a packet to any registered role that advertises the given capability.
     * Use strategy to control which role is picked when multiple qualify.
     *
     * @example
     * const result = await runtime.sendToCapability("summarize", packet);
     * const result = await runtime.sendToCapability("translate", packet, "least-loaded");
     */
    async sendToCapability(capability, packet, strategy = "round-robin") {
        const candidates = this.registry
            .findByCapability(capability)
            .map((e) => e.registration.name);
        if (candidates.length === 0) {
            this.stats.packetsFailed++;
            return this.buildErrorPacket(packet, errors_1.ErrorCode.NO_CAPABLE_ROLE, `No registered role has capability "${capability}"`);
        }
        const chosen = this.loadBalancer.pick(candidates, strategy);
        const routed = { ...packet, target: chosen };
        this.loadBalancer.recordStart(chosen);
        try {
            const reply = await this.send(routed);
            return reply;
        }
        finally {
            this.loadBalancer.recordEnd(chosen);
        }
    }
    // ── Item 7: Ack/nack ──────────────────────
    /**
     * Send a packet and receive two separate promises:
     *   - ack:    resolves when the handler acknowledges receipt (task.ack)
     *   - result: resolves when the handler calls runtime.deliverResult()
     *
     * The handler should return task.ack immediately and later call:
     *   runtime.deliverResult(packet.taskId, resultPacket)
     *
     * @example
     * const { ack, result } = await runtime.sendWithAck(packet);
     * console.log("acknowledged:", ack.type);  // "task.ack"
     * const final = await result;              // arrives asynchronously
     */
    async sendWithAck(packet) {
        const resultPromise = new Promise((resolve, reject) => {
            this.pendingAcks.set(packet.taskId, { resolve, reject });
        });
        const augmented = {
            ...packet,
            meta: { ...packet.meta, ackRequired: true },
        };
        const ack = await this.send(augmented);
        // If the send itself failed, reject the result promise
        if (ack.type === "error.report" || ack.type === "task.nack") {
            const pending = this.pendingAcks.get(packet.taskId);
            if (pending) {
                pending.reject(new errors_1.MailmanError(errors_1.ErrorCode.HANDLER_THREW, ack.error?.message ?? "Handler rejected the packet"));
                this.pendingAcks.delete(packet.taskId);
            }
        }
        return { ack, result: resultPromise };
    }
    /**
     * Deliver the final result for a task that was sent with sendWithAck().
     * Call this from inside the handler after async processing is done.
     *
     * @param taskId  The original packet's taskId.
     * @param result  The final result packet to resolve the pending promise.
     */
    deliverResult(taskId, result) {
        const pending = this.pendingAcks.get(taskId);
        if (!pending)
            return;
        pending.resolve(result);
        this.pendingAcks.delete(taskId);
        this.record(result.packetId, taskId, "packet.completed", "mailman", {
            via: "deliverResult",
        });
    }
    // ── Item 8: Scheduling ────────────────────
    /**
     * Schedule a packet for delivery at a specific date/time.
     * Returns a scheduleId for cancellation.
     */
    schedulePacket(packet, deliverAt) {
        const id = this.scheduler.schedule(packet, deliverAt);
        this.record(packet.packetId, packet.taskId, "packet.scheduled", "mailman", {
            scheduleId: id,
            deliverAt: deliverAt.toISOString(),
        });
        return id;
    }
    /**
     * Schedule a packet for delivery after a delay.
     * Returns a scheduleId for cancellation.
     */
    scheduleAfter(packet, delayMs) {
        const id = this.scheduler.scheduleAfter(packet, delayMs);
        this.record(packet.packetId, packet.taskId, "packet.scheduled", "mailman", {
            scheduleId: id,
            delayMs,
        });
        return id;
    }
    /**
     * Schedule a packet to be delivered repeatedly.
     * Returns a scheduleId — pass to cancelScheduled() to stop.
     */
    scheduleRecurring(packet, intervalMs, startAt) {
        const id = this.scheduler.scheduleRecurring(packet, intervalMs, startAt);
        this.record(packet.packetId, packet.taskId, "packet.scheduled", "mailman", {
            scheduleId: id,
            intervalMs,
            recurring: true,
        });
        return id;
    }
    /** Cancel a scheduled or recurring packet. Returns true if found. */
    cancelScheduled(scheduleId) {
        return this.scheduler.cancel(scheduleId);
    }
    /** All packets currently waiting to be delivered. */
    listScheduled() {
        return this.scheduler.listPending();
    }
    // ── Ping ─────────────────────────────────
    async ping(roleName) {
        const entry = this.registry.get(roleName);
        if (!entry)
            return false;
        if (!entry.registration.accepts.includes("health.ping"))
            return false;
        const pingPacket = {
            packetId: (0, ids_1.newPacketId)(),
            taskId: "ping",
            type: "health.ping",
            sender: "mailman",
            target: roleName,
            timestamp: (0, time_1.now)(),
            payload: {},
        };
        try {
            const pong = await this.router.dispatch(pingPacket);
            return pong.type === "health.pong";
        }
        catch {
            return false;
        }
    }
    // ── Trace access ──────────────────────────
    getTrace(packetId) {
        return this.traceStore.getByPacket(packetId);
    }
    getTaskTrace(taskId) {
        return this.traceStore.getByTask(taskId);
    }
    allTraces() {
        return this.traceStore.all();
    }
    // ── DLQ access ────────────────────────────
    getDLQ() {
        return this.dlq;
    }
    // ── Stats ─────────────────────────────────
    getStats() {
        return { ...this.stats };
    }
    // ── Internal helpers ──────────────────────
    /** Unified trace record — calls traceStore and fires telemetry hook. */
    record(packetId, taskId, event, actor, details) {
        this.traceStore.record(packetId, taskId, event, actor, details);
        if (this.telemetryHook) {
            try {
                this.telemetryHook({ type: event, packetId, taskId, actor, timestamp: (0, time_1.now)(), details });
            }
            catch {
                // hooks must not crash the runtime
            }
        }
    }
    buildErrorPacket(original, code, message, details) {
        return {
            packetId: (0, ids_1.newPacketId)(),
            taskId: original.taskId,
            parentPacketId: original.packetId,
            type: "error.report",
            sender: "mailman",
            target: original.sender,
            timestamp: (0, time_1.now)(),
            payload: {},
            status: "failed",
            error: { code, message, details },
        };
    }
    withTimeout(promise, ms, original) {
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                reject(new errors_1.MailmanError(errors_1.ErrorCode.HANDLER_TIMEOUT, `Handler for "${original.target}" timed out after ${ms}ms`));
            }, ms);
            promise
                .then((r) => { clearTimeout(timer); resolve(r); })
                .catch((e) => { clearTimeout(timer); reject(e); });
        });
    }
}
exports.Runtime = Runtime;
// ─────────────────────────────────────────────
//  Singleton (for CLI / simple usage)
// ─────────────────────────────────────────────
let _instance = null;
function getRuntime(config) {
    if (!_instance) {
        _instance = new Runtime(config);
    }
    return _instance;
}
function resetRuntime() {
    _instance = null;
}
