// @marsulta/mailman/dist/core/scheduler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PacketScheduler = void 0;
const ids_1 = require("../utils/ids");
const time_1 = require("../utils/time");
// ─────────────────────────────────────────────
//  PacketScheduler
//
//  In-memory scheduler backed by setTimeout.
//  Schedules survive cancel() calls but NOT
//  process restarts — for persistence, pair
//  with the scheduled_packets SQLite table.
// ─────────────────────────────────────────────
class PacketScheduler {
    dispatch;
    pending = new Map();
    constructor(dispatch) {
        this.dispatch = dispatch;
    }
    /**
     * Deliver packet once at the given date.
     * Returns the scheduleId.
     */
    schedule(packet, deliverAt) {
        const entry = {
            scheduleId: (0, ids_1.newId)("sched"),
            packet,
            deliverAt: deliverAt.getTime(),
            createdAt: (0, time_1.now)(),
        };
        this.arm(entry, Math.max(0, deliverAt.getTime() - Date.now()));
        return entry.scheduleId;
    }
    /**
     * Deliver packet once after delayMs milliseconds.
     * Returns the scheduleId.
     */
    scheduleAfter(packet, delayMs) {
        return this.schedule(packet, new Date(Date.now() + delayMs));
    }
    /**
     * Deliver packet repeatedly every intervalMs milliseconds.
     * First delivery is after intervalMs (or at startAt if provided).
     * Returns the scheduleId — pass to cancel() to stop.
     */
    scheduleRecurring(packet, intervalMs, startAt) {
        const first = startAt ?? new Date(Date.now() + intervalMs);
        const entry = {
            scheduleId: (0, ids_1.newId)("sched"),
            packet,
            deliverAt: first.getTime(),
            recurring: { intervalMs },
            createdAt: (0, time_1.now)(),
        };
        this.arm(entry, Math.max(0, first.getTime() - Date.now()));
        return entry.scheduleId;
    }
    /**
     * Cancel a pending schedule.
     * Returns true if found and cancelled, false if not found.
     */
    cancel(scheduleId) {
        const item = this.pending.get(scheduleId);
        if (!item)
            return false;
        clearTimeout(item.timer);
        this.pending.delete(scheduleId);
        return true;
    }
    /** All schedules currently waiting to fire. */
    listPending() {
        return Array.from(this.pending.values()).map((v) => v.entry);
    }
    // ── Internal ──────────────────────────────
    arm(entry, delayMs) {
        const timer = setTimeout(() => {
            this.pending.delete(entry.scheduleId);
            this.dispatch(entry.packet);
            if (entry.recurring) {
                const next = {
                    ...entry,
                    deliverAt: Date.now() + entry.recurring.intervalMs,
                };
                this.arm(next, entry.recurring.intervalMs);
            }
        }, delayMs);
        // Prevent long-running timers from keeping the process alive
        if (typeof timer === "object" && "unref" in timer) {
            timer.unref();
        }
        this.pending.set(entry.scheduleId, { entry, timer });
    }
}
exports.PacketScheduler = PacketScheduler;
