// @marsulta/mailman/dist/core/streamSession.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamSessionStore = exports.StreamSession = void 0;
class StreamSession {
    queue = [];
    waiters = [];
    _ended = false;
    get isEnded() {
        return this._ended;
    }
    // ── Producer API (handler side) ───────────
    push(chunk) {
        if (this._ended)
            return;
        this.enqueue({ kind: "chunk", value: chunk });
    }
    end() {
        if (this._ended)
            return;
        this._ended = true;
        this.enqueue({ kind: "end" });
    }
    error(err) {
        if (this._ended)
            return;
        this._ended = true;
        this.enqueue({ kind: "error", error: err });
    }
    // ── Consumer API (async iterable) ─────────
    async *[Symbol.asyncIterator]() {
        while (true) {
            const chunk = await this.dequeue();
            if (chunk.kind === "end")
                return;
            if (chunk.kind === "error")
                throw chunk.error;
            yield chunk.value;
        }
    }
    // ── Internal plumbing ─────────────────────
    enqueue(chunk) {
        const waiter = this.waiters.shift();
        if (waiter) {
            waiter(chunk);
        }
        else {
            this.queue.push(chunk);
        }
    }
    dequeue() {
        const queued = this.queue.shift();
        if (queued)
            return Promise.resolve(queued);
        return new Promise((resolve) => this.waiters.push(resolve));
    }
}
exports.StreamSession = StreamSession;
// ─────────────────────────────────────────────
//  StreamSessionStore — tracks all active sessions
// ─────────────────────────────────────────────
class StreamSessionStore {
    sessions = new Map();
    create(streamId) {
        const session = new StreamSession();
        this.sessions.set(streamId, session);
        return session;
    }
    get(streamId) {
        return this.sessions.get(streamId);
    }
    delete(streamId) {
        this.sessions.delete(streamId);
    }
    has(streamId) {
        return this.sessions.has(streamId);
    }
    activeCount() {
        return this.sessions.size;
    }
}
exports.StreamSessionStore = StreamSessionStore;
