// @marsulta/mailman/dist/core/traceStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TraceStore = void 0;
const ids_1 = require("../utils/ids");
const time_1 = require("../utils/time");
const db_1 = require("./db");
// ─────────────────────────────────────────────
//  TraceStore — SQLite-backed, persistent
// ─────────────────────────────────────────────
class TraceStore {
    db;
    /**
     * @param dbPath SQLite file path. Defaults to ~/.mailman/traces.db.
     *               Pass ':memory:' for ephemeral/test usage.
     */
    constructor(dbPath) {
        this.db = (0, db_1.openDb)(dbPath);
    }
    // ── Write ──────────────────────────────────
    record(packetId, taskId, event, actor, details) {
        const entry = {
            traceId: (0, ids_1.newId)("trace"),
            packetId,
            taskId,
            event,
            actor,
            timestamp: (0, time_1.now)(),
            details,
        };
        this.db
            .prepare(`INSERT INTO trace_events
         (trace_id, packet_id, task_id, event, actor, timestamp, details)
         VALUES (?, ?, ?, ?, ?, ?, ?)`)
            .run(entry.traceId, packetId, taskId, event, actor, entry.timestamp, details !== undefined ? JSON.stringify(details) : null);
        return entry;
    }
    // ── Read ───────────────────────────────────
    getByPacket(packetId) {
        return this.db
            .prepare("SELECT * FROM trace_events WHERE packet_id = ? ORDER BY timestamp ASC")
            .all(packetId).map(rowToEntry);
    }
    getByTask(taskId) {
        return this.db
            .prepare("SELECT * FROM trace_events WHERE task_id = ? ORDER BY timestamp ASC")
            .all(taskId).map(rowToEntry);
    }
    all() {
        return this.db
            .prepare("SELECT * FROM trace_events ORDER BY timestamp ASC")
            .all().map(rowToEntry);
    }
    totalEvents() {
        const row = this.db
            .prepare("SELECT COUNT(*) as count FROM trace_events")
            .get();
        return row.count;
    }
    // ── Query (for Postmaster / CLI) ───────────
    /**
     * Flexible filter query — used by Postmaster and `mailman inspect`.
     */
    query({ since, until, actor, event, limit = 100, }) {
        const conditions = [];
        const params = [];
        if (since) {
            conditions.push("timestamp >= ?");
            params.push(since);
        }
        if (until) {
            conditions.push("timestamp <= ?");
            params.push(until);
        }
        if (actor) {
            conditions.push("actor = ?");
            params.push(actor);
        }
        if (event) {
            conditions.push("event = ?");
            params.push(event);
        }
        const where = conditions.length > 0 ? `WHERE ${conditions.join(" AND ")}` : "";
        const sql = `SELECT * FROM trace_events ${where} ORDER BY timestamp DESC LIMIT ?`;
        params.push(limit);
        return this.db.prepare(sql).all(...params).map(rowToEntry);
    }
    // ── Maintenance ────────────────────────────
    clear() {
        this.db.exec("DELETE FROM trace_events");
    }
}
exports.TraceStore = TraceStore;
// ─────────────────────────────────────────────
//  Row → domain type
// ─────────────────────────────────────────────
function rowToEntry(row) {
    return {
        traceId: row["trace_id"],
        packetId: row["packet_id"],
        taskId: row["task_id"],
        event: row["event"],
        actor: row["actor"],
        timestamp: row["timestamp"],
        details: typeof row["details"] === "string"
            ? JSON.parse(row["details"])
            : undefined,
    };
}
