// @marsulta/mailman/dist/core/validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Validator = void 0;
exports.assertValid = assertValid;
const types_1 = require("../packet/types");
const errors_1 = require("./errors");
// ─────────────────────────────────────────────
//  Validator
// ─────────────────────────────────────────────
class Validator {
    registry;
    customTypes;
    /**
     * @param registry   Role registry — used to verify target exists and accepts the type.
     * @param customTypes Extra packet types registered by the user via runtime.registerPacketType().
     */
    constructor(registry, customTypes = new Set()) {
        this.registry = registry;
        this.customTypes = customTypes;
    }
    validate(packet) {
        // Required string fields
        const requiredStrings = [
            "packetId",
            "taskId",
            "type",
            "sender",
            "target",
        ];
        for (const field of requiredStrings) {
            const value = packet[field];
            if (value === undefined || value === null || value === "") {
                return {
                    ok: false,
                    code: errors_1.ErrorCode.INVALID_PACKET,
                    message: `Missing required field: ${field}`,
                };
            }
        }
        // payload must exist (can be empty object)
        if (packet.payload === undefined || packet.payload === null) {
            return {
                ok: false,
                code: errors_1.ErrorCode.INVALID_PACKET,
                message: "Missing required field: payload",
            };
        }
        // type must be a known built-in OR a user-registered custom type
        const isBuiltin = types_1.BUILTIN_PACKET_TYPES.includes(packet.type);
        const isCustom = this.customTypes.has(packet.type);
        if (!isBuiltin && !isCustom) {
            return {
                ok: false,
                code: errors_1.ErrorCode.INVALID_PACKET,
                message: `Unknown packet type: "${packet.type}". Register custom types with runtime.registerPacketType()`,
            };
        }
        // confidence must be 0–1 if present
        if (packet.confidence !== undefined) {
            if (typeof packet.confidence !== "number" ||
                packet.confidence < 0 ||
                packet.confidence > 1) {
                return {
                    ok: false,
                    code: errors_1.ErrorCode.INVALID_PACKET,
                    message: `confidence must be a number between 0 and 1 (got ${packet.confidence})`,
                };
            }
        }
        // target must be registered
        const role = this.registry.get(packet.target);
        if (!role) {
            return {
                ok: false,
                code: errors_1.ErrorCode.UNKNOWN_TARGET,
                message: `Target not registered: ${packet.target}`,
            };
        }
        // type must be accepted by target
        if (!role.registration.accepts.includes(packet.type)) {
            return {
                ok: false,
                code: errors_1.ErrorCode.TARGET_REJECTS_TYPE,
                message: `Target "${packet.target}" does not accept type "${packet.type}"`,
            };
        }
        return { ok: true };
    }
}
exports.Validator = Validator;
/**
 * Convenience assertion — throws MailmanError on failure.
 */
function assertValid(packet, registry, customTypes) {
    const result = new Validator(registry, customTypes).validate(packet);
    if (!result.ok) {
        throw new errors_1.MailmanError(result.code, result.message);
    }
}
