// @marsulta/mailman/dist/core/__tests__/features.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const runtime_1 = require("../runtime");
const backends_1 = require("../backends");
const ids_1 = require("../../utils/ids");
const time_1 = require("../../utils/time");
// ─────────────────────────────────────────────
//  Helpers
// ─────────────────────────────────────────────
function makeRuntime() {
    return new runtime_1.Runtime({
        traceStore: new backends_1.MemoryTraceStore(),
        dlq: new backends_1.MemoryDLQStore(),
    });
}
function pkt(overrides = {}) {
    return {
        packetId: (0, ids_1.newPacketId)(),
        taskId: (0, ids_1.newTaskId)(),
        type: "task.assign",
        sender: "test",
        target: "worker",
        timestamp: (0, time_1.now)(),
        payload: {},
        ...overrides,
    };
}
// ─────────────────────────────────────────────
//  Item 1 — Custom packet types
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Item 1: custom packet types", () => {
    (0, vitest_1.it)("rejects unknown type before registration", async () => {
        const rt = makeRuntime();
        rt.start();
        rt.registerRole({ name: "worker", accepts: ["my.event"] }, async (p) => p);
        const reply = await rt.send(pkt({ type: "my.event" }));
        (0, vitest_1.expect)(reply.type).toBe("error.report");
        (0, vitest_1.expect)(reply.error?.message).toMatch(/Unknown packet type/);
    });
    (0, vitest_1.it)("accepts type after registerPacketType()", async () => {
        const rt = makeRuntime();
        rt.start();
        rt.registerPacketType("my.event");
        rt.registerRole({ name: "worker", accepts: ["my.event"] }, async (p) => ({
            ...p, type: "task.result", sender: "worker", target: p.sender,
        }));
        const reply = await rt.send(pkt({ type: "my.event" }));
        (0, vitest_1.expect)(reply.type).toBe("task.result");
    });
});
// ─────────────────────────────────────────────
//  Item 2 — Pub/Sub + fire-and-forget
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Item 2: pub/sub + fire-and-forget", () => {
    (0, vitest_1.it)("publish delivers to exact-match subscriber", () => {
        const rt = makeRuntime();
        rt.start();
        const received = [];
        rt.subscribe("task.result", (p) => received.push(p));
        const packet = pkt({ type: "task.result" });
        rt.publish(packet);
        (0, vitest_1.expect)(received).toHaveLength(1);
        (0, vitest_1.expect)(received[0].packetId).toBe(packet.packetId);
    });
    (0, vitest_1.it)("wildcard topic matches prefix", () => {
        const rt = makeRuntime();
        rt.start();
        const received = [];
        rt.subscribe("task.*", (p) => received.push(p));
        rt.publish(pkt({ type: "task.assign" }));
        rt.publish(pkt({ type: "task.result" }));
        rt.publish(pkt({ type: "health.ping" })); // should NOT match
        (0, vitest_1.expect)(received).toHaveLength(2);
    });
    (0, vitest_1.it)("unsubscribe stops delivery", () => {
        const rt = makeRuntime();
        rt.start();
        const received = [];
        const unsub = rt.subscribe("*", (p) => received.push(p));
        rt.publish(pkt({ type: "task.assign" }));
        unsub();
        rt.publish(pkt({ type: "task.assign" }));
        (0, vitest_1.expect)(received).toHaveLength(1);
    });
    (0, vitest_1.it)("dispatch is fire-and-forget (returns void)", async () => {
        const rt = makeRuntime();
        rt.start();
        const handled = [];
        rt.registerRole({ name: "worker", accepts: ["task.assign"] }, async (p) => {
            handled.push(p.packetId);
            return { ...p, type: "task.result", sender: "worker", target: p.sender };
        });
        rt.dispatch(pkt());
        // give the async work a tick
        await new Promise((r) => setTimeout(r, 20));
        (0, vitest_1.expect)(handled).toHaveLength(1);
    });
});
// ─────────────────────────────────────────────
//  Item 4 — Streaming
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Item 4: streaming", () => {
    (0, vitest_1.it)("streams chunks to consumer", async () => {
        const rt = makeRuntime();
        rt.start();
        rt.registerStreamRole({ name: "worker", accepts: ["stream.open"] }, async (_packet, stream) => {
            stream.push("hello");
            stream.push(" ");
            stream.push("world");
            stream.end();
        });
        const iterable = await rt.openStream(pkt({ type: "stream.open", target: "worker" }));
        const chunks = [];
        for await (const chunk of iterable) {
            chunks.push(chunk);
        }
        (0, vitest_1.expect)(chunks.join("")).toBe("hello world");
    });
    (0, vitest_1.it)("propagates stream errors", async () => {
        const rt = makeRuntime();
        rt.start();
        rt.registerStreamRole({ name: "worker", accepts: ["stream.open"] }, async (_packet, stream) => {
            stream.push("partial");
            stream.error(new Error("boom"));
        });
        const iterable = await rt.openStream(pkt({ type: "stream.open", target: "worker" }));
        const chunks = [];
        await (0, vitest_1.expect)(async () => {
            for await (const chunk of iterable) {
                chunks.push(chunk);
            }
        }).rejects.toThrow("boom");
        (0, vitest_1.expect)(chunks).toEqual(["partial"]);
    });
});
// ─────────────────────────────────────────────
//  Item 6 — Capability routing + load balancing
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Item 6: capability routing", () => {
    (0, vitest_1.it)("routes to a role with the requested capability", async () => {
        const rt = makeRuntime();
        rt.start();
        const handled = [];
        rt.registerRole({ name: "summarizer", accepts: ["task.assign"], capabilities: ["summarize"] }, async (p) => {
            handled.push("summarizer");
            return { ...p, type: "task.result", sender: "summarizer", target: p.sender };
        });
        await rt.sendToCapability("summarize", pkt());
        (0, vitest_1.expect)(handled).toContain("summarizer");
    });
    (0, vitest_1.it)("returns error packet when no capable role exists", async () => {
        const rt = makeRuntime();
        rt.start();
        const reply = await rt.sendToCapability("translate", pkt());
        (0, vitest_1.expect)(reply.type).toBe("error.report");
        (0, vitest_1.expect)(reply.error?.code).toBe("NO_CAPABLE_ROLE");
    });
    (0, vitest_1.it)("round-robins across multiple capable roles", async () => {
        const rt = makeRuntime();
        rt.start();
        const hits = [];
        for (const name of ["w1", "w2", "w3"]) {
            const n = name;
            rt.registerRole({ name: n, accepts: ["task.assign"], capabilities: ["work"] }, async (p) => {
                hits.push(n);
                return { ...p, type: "task.result", sender: n, target: p.sender };
            });
        }
        await rt.sendToCapability("work", pkt());
        await rt.sendToCapability("work", pkt());
        await rt.sendToCapability("work", pkt());
        // All three should have been called (round-robin)
        (0, vitest_1.expect)(new Set(hits).size).toBe(3);
    });
});
// ─────────────────────────────────────────────
//  Item 7 — Ack/nack
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Item 7: ack/nack", () => {
    (0, vitest_1.it)("sendWithAck receives ack immediately then result asynchronously", async () => {
        const rt = makeRuntime();
        rt.start();
        rt.registerRole({ name: "worker", accepts: ["task.assign"] }, async (p) => {
            // Simulate async work in background
            setTimeout(() => {
                rt.deliverResult(p.taskId, {
                    ...p,
                    packetId: (0, ids_1.newPacketId)(),
                    type: "task.result",
                    sender: "worker",
                    target: p.sender,
                    payload: { done: true },
                });
            }, 10);
            // Return immediate ack
            return { ...p, packetId: (0, ids_1.newPacketId)(), type: "task.ack", sender: "worker", target: p.sender };
        });
        const { ack, result } = await rt.sendWithAck(pkt());
        (0, vitest_1.expect)(ack.type).toBe("task.ack");
        const final = await result;
        (0, vitest_1.expect)(final.type).toBe("task.result");
        (0, vitest_1.expect)(final.payload.done).toBe(true);
    });
});
// ─────────────────────────────────────────────
//  Item 8 — Scheduling
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Item 8: scheduling", () => {
    (0, vitest_1.it)("scheduleAfter delivers packet after delay", async () => {
        const rt = makeRuntime();
        rt.start();
        const received = [];
        rt.registerRole({ name: "worker", accepts: ["task.assign"] }, async (p) => {
            received.push(p);
            return { ...p, type: "task.result", sender: "worker", target: p.sender };
        });
        (0, vitest_1.expect)(received).toHaveLength(0);
        rt.scheduleAfter(pkt(), 20);
        await new Promise((r) => setTimeout(r, 50));
        (0, vitest_1.expect)(received).toHaveLength(1);
    });
    (0, vitest_1.it)("cancel prevents delivery", async () => {
        const rt = makeRuntime();
        rt.start();
        const received = [];
        rt.registerRole({ name: "worker", accepts: ["task.assign"] }, async (p) => {
            received.push(p);
            return { ...p, type: "task.result", sender: "worker", target: p.sender };
        });
        const id = rt.scheduleAfter(pkt(), 30);
        const cancelled = rt.cancelScheduled(id);
        await new Promise((r) => setTimeout(r, 60));
        (0, vitest_1.expect)(cancelled).toBe(true);
        (0, vitest_1.expect)(received).toHaveLength(0);
    });
});
// ─────────────────────────────────────────────
//  Item 9 — Telemetry hooks
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Item 9: telemetry hooks", () => {
    (0, vitest_1.it)("onTelemetry fires for each trace event", async () => {
        const events = [];
        const rt = new runtime_1.Runtime({
            traceStore: new backends_1.MemoryTraceStore(),
            dlq: new backends_1.MemoryDLQStore(),
            onTelemetry: (e) => events.push(e.type),
        });
        rt.start();
        rt.registerRole({ name: "worker", accepts: ["task.assign"] }, async (p) => ({
            ...p, type: "task.result", sender: "worker", target: p.sender,
        }));
        await rt.send(pkt());
        (0, vitest_1.expect)(events).toContain("packet.received");
        (0, vitest_1.expect)(events).toContain("packet.validated");
        (0, vitest_1.expect)(events).toContain("packet.completed");
    });
});
// ─────────────────────────────────────────────
//  Item 10 — Pluggable backends
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Item 10: pluggable backends", () => {
    (0, vitest_1.it)("MemoryTraceStore records and retrieves entries", async () => {
        const store = new backends_1.MemoryTraceStore();
        const rt = new runtime_1.Runtime({ traceStore: store, dlq: new backends_1.MemoryDLQStore() });
        rt.start();
        rt.registerRole({ name: "worker", accepts: ["task.assign"] }, async (p) => ({
            ...p, type: "task.result", sender: "worker", target: p.sender,
        }));
        const packet = pkt();
        await rt.send(packet);
        const trace = store.getByPacket(packet.packetId);
        (0, vitest_1.expect)(trace.length).toBeGreaterThan(0);
        (0, vitest_1.expect)(trace[0].event).toBe("packet.received");
    });
    (0, vitest_1.it)("MemoryDLQStore captures dead-lettered packets", async () => {
        const dlq = new backends_1.MemoryDLQStore();
        const rt = new runtime_1.Runtime({
            traceStore: new backends_1.MemoryTraceStore(),
            dlq,
            retry: { maxAttempts: 1, baseDelayMs: 0, maxDelayMs: 0 },
        });
        rt.start();
        rt.registerRole({ name: "worker", accepts: ["task.assign"] }, async () => {
            throw new Error("intentional failure");
        });
        await rt.send(pkt());
        (0, vitest_1.expect)(dlq.count()).toBe(1);
    });
});
