// @marsulta/mailman/dist/core/__tests__/runtime.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const runtime_1 = require("../runtime");
const createPacket_1 = require("../../packet/createPacket");
// ─────────────────────────────────────────────
//  Helpers
// ─────────────────────────────────────────────
function makeRuntime() {
    const rt = new runtime_1.Runtime({ dbPath: ":memory:", retry: { maxAttempts: 1 } });
    rt.start();
    return rt;
}
function echoHandler(packet) {
    return Promise.resolve((0, createPacket_1.createPacket)({
        taskId: packet.taskId,
        parentPacketId: packet.packetId,
        type: "task.result",
        sender: packet.target,
        target: packet.sender,
        payload: { echo: packet.payload },
        status: "completed",
    }));
}
// ─────────────────────────────────────────────
//  Tests
// ─────────────────────────────────────────────
(0, vitest_1.describe)("Runtime", () => {
    let rt;
    (0, vitest_1.beforeEach)(() => {
        rt = makeRuntime();
    });
    // ── Happy path ──────────────────────────────
    (0, vitest_1.it)("delivers a packet and returns a response", async () => {
        rt.registerRole({ name: "worker", accepts: ["task.assign"] }, (pkt) => Promise.resolve((0, createPacket_1.createPacket)({
            taskId: pkt.taskId,
            type: "task.result",
            sender: "worker",
            target: pkt.sender,
            payload: { done: true },
            status: "completed",
        })));
        const packet = (0, createPacket_1.createPacket)({
            taskId: "t1",
            type: "task.assign",
            sender: "orchestrator",
            target: "worker",
            payload: { prompt: "hello" },
        });
        const reply = await rt.send(packet);
        (0, vitest_1.expect)(reply.type).toBe("task.result");
        (0, vitest_1.expect)(reply.status).toBe("completed");
    });
    // ── Validation rejection ─────────────────────
    (0, vitest_1.it)("returns an error packet when target is not registered", async () => {
        const packet = (0, createPacket_1.createPacket)({
            taskId: "t2",
            type: "task.assign",
            sender: "orchestrator",
            target: "ghost",
            payload: {},
        });
        const reply = await rt.send(packet);
        (0, vitest_1.expect)(reply.type).toBe("error.report");
        (0, vitest_1.expect)(reply.error?.code).toBe("UNKNOWN_TARGET");
    });
    (0, vitest_1.it)("returns an error packet when required field is missing", async () => {
        const packet = {
            packetId: "p1",
            taskId: "t3",
            type: "task.assign",
            sender: "",
            target: "worker",
            timestamp: new Date().toISOString(),
            payload: {},
        };
        const reply = await rt.send(packet);
        (0, vitest_1.expect)(reply.type).toBe("error.report");
        (0, vitest_1.expect)(reply.error?.message).toMatch(/sender/i);
    });
    // ── Handler error ────────────────────────────
    (0, vitest_1.it)("returns an error packet when handler throws", async () => {
        rt.registerRole({ name: "crasher", accepts: ["task.assign"] }, () => {
            throw new Error("boom");
        });
        const packet = (0, createPacket_1.createPacket)({
            taskId: "t4",
            type: "task.assign",
            sender: "orchestrator",
            target: "crasher",
            payload: {},
        });
        const reply = await rt.send(packet);
        (0, vitest_1.expect)(reply.type).toBe("error.report");
        (0, vitest_1.expect)(reply.error?.message).toBe("boom");
    });
    // ── Timeout ──────────────────────────────────
    (0, vitest_1.it)("times out a slow handler", async () => {
        rt.registerRole({ name: "slow", accepts: ["task.assign"] }, () => new Promise((resolve) => setTimeout(resolve, 500)));
        const packet = (0, createPacket_1.createPacket)({
            taskId: "t5",
            type: "task.assign",
            sender: "orchestrator",
            target: "slow",
            payload: {},
            meta: { timeoutMs: 50 },
        });
        const reply = await rt.send(packet);
        (0, vitest_1.expect)(reply.type).toBe("error.report");
        (0, vitest_1.expect)(reply.error?.code).toBe("HANDLER_TIMEOUT");
    }, 10_000);
    // ── Middleware ───────────────────────────────
    (0, vitest_1.it)("runs middleware in order, wrapping the handler", async () => {
        const log = [];
        rt.registerRole({ name: "target", accepts: ["task.assign"] }, echoHandler);
        rt.use(async (pkt, next) => {
            log.push("mw1-before");
            const reply = await next();
            log.push("mw1-after");
            return reply;
        });
        rt.use(async (pkt, next) => {
            log.push("mw2-before");
            const reply = await next();
            log.push("mw2-after");
            return reply;
        });
        const packet = (0, createPacket_1.createPacket)({
            taskId: "t6",
            type: "task.assign",
            sender: "orch",
            target: "target",
            payload: {},
        });
        await rt.send(packet);
        (0, vitest_1.expect)(log).toEqual(["mw1-before", "mw2-before", "mw2-after", "mw1-after"]);
    });
    // ── Ping ─────────────────────────────────────
    (0, vitest_1.it)("ping returns true when role accepts health.ping", async () => {
        rt.registerRole({ name: "pinger", accepts: ["health.ping"] }, (pkt) => Promise.resolve((0, createPacket_1.createPacket)({
            taskId: pkt.taskId,
            type: "health.pong",
            sender: "pinger",
            target: pkt.sender,
            payload: {},
        })));
        (0, vitest_1.expect)(await rt.ping("pinger")).toBe(true);
    });
    (0, vitest_1.it)("ping returns false for unknown role", async () => {
        (0, vitest_1.expect)(await rt.ping("nobody")).toBe(false);
    });
    // ── Trace ─────────────────────────────────────
    (0, vitest_1.it)("records trace events for a successful send", async () => {
        rt.registerRole({ name: "tracer-target", accepts: ["task.assign"] }, echoHandler);
        const packet = (0, createPacket_1.createPacket)({
            taskId: "t7",
            type: "task.assign",
            sender: "orch",
            target: "tracer-target",
            payload: {},
        });
        await rt.send(packet);
        const trace = rt.getTrace(packet.packetId);
        const events = trace.map((e) => e.event);
        (0, vitest_1.expect)(events).toContain("packet.received");
        (0, vitest_1.expect)(events).toContain("packet.validated");
        (0, vitest_1.expect)(events).toContain("handler.started");
        (0, vitest_1.expect)(events).toContain("packet.completed");
    });
    // ── Stats ─────────────────────────────────────
    (0, vitest_1.it)("increments packetsProcessed on success", async () => {
        rt.registerRole({ name: "counter", accepts: ["task.assign"] }, echoHandler);
        const packet = (0, createPacket_1.createPacket)({
            taskId: "t8",
            type: "task.assign",
            sender: "orch",
            target: "counter",
            payload: {},
        });
        await rt.send(packet);
        (0, vitest_1.expect)(rt.getStats().packetsProcessed).toBe(1);
    });
    // ── DLQ ───────────────────────────────────────
    (0, vitest_1.it)("pushes a packet to the DLQ after handler exhausts retries", async () => {
        // Runtime with maxAttempts:1 so it fails immediately into DLQ
        const rt2 = new runtime_1.Runtime({
            dbPath: ":memory:",
            retry: { maxAttempts: 1 },
        });
        rt2.start();
        rt2.registerRole({ name: "boom", accepts: ["task.assign"] }, () => {
            throw new Error("always fails");
        });
        const packet = (0, createPacket_1.createPacket)({
            taskId: "dlq-test",
            type: "task.assign",
            sender: "orch",
            target: "boom",
            payload: {},
        });
        await rt2.send(packet);
        const dlq = rt2.getDLQ();
        (0, vitest_1.expect)(dlq.count()).toBe(1);
        const entries = dlq.list();
        (0, vitest_1.expect)(entries[0].error).toBe("always fails");
        (0, vitest_1.expect)(entries[0].attempts).toBe(1);
    });
});
