// @marsulta/mailman/dist/core/__tests__/validator.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const validator_1 = require("../validator");
const registry_1 = require("../registry");
function validPacket(overrides = {}) {
    return {
        packetId: "pkt-1",
        taskId: "task-1",
        type: "task.assign",
        sender: "orchestrator",
        target: "worker",
        timestamp: new Date().toISOString(),
        payload: { prompt: "do work" },
        ...overrides,
    };
}
(0, vitest_1.describe)("Validator", () => {
    let registry;
    let validator;
    (0, vitest_1.beforeEach)(() => {
        registry = new registry_1.Registry();
        registry.register({ name: "worker", accepts: ["task.assign", "health.ping"] }, async (p) => p);
        validator = new validator_1.Validator(registry);
    });
    (0, vitest_1.it)("passes a valid packet", () => {
        const result = validator.validate(validPacket());
        (0, vitest_1.expect)(result.ok).toBe(true);
    });
    (0, vitest_1.it)("rejects missing packetId", () => {
        const result = validator.validate(validPacket({ packetId: "" }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.message).toMatch(/packetId/i);
    });
    (0, vitest_1.it)("rejects missing taskId", () => {
        const result = validator.validate(validPacket({ taskId: "" }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.message).toMatch(/taskId/i);
    });
    (0, vitest_1.it)("rejects missing sender", () => {
        const result = validator.validate(validPacket({ sender: "" }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.message).toMatch(/sender/i);
    });
    (0, vitest_1.it)("rejects missing target", () => {
        const result = validator.validate(validPacket({ target: "" }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.message).toMatch(/target/i);
    });
    (0, vitest_1.it)("rejects null payload", () => {
        const result = validator.validate(validPacket({ payload: undefined }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.message).toMatch(/payload/i);
    });
    (0, vitest_1.it)("rejects unknown packet type", () => {
        const result = validator.validate(validPacket({ type: "unknown.type" }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.message).toMatch(/Unknown packet type/i);
    });
    (0, vitest_1.it)("rejects confidence below 0", () => {
        const result = validator.validate(validPacket({ confidence: -0.1 }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.message).toMatch(/confidence/i);
    });
    (0, vitest_1.it)("rejects confidence above 1", () => {
        const result = validator.validate(validPacket({ confidence: 1.5 }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.message).toMatch(/confidence/i);
    });
    (0, vitest_1.it)("accepts confidence exactly at boundaries (0 and 1)", () => {
        (0, vitest_1.expect)(validator.validate(validPacket({ confidence: 0 })).ok).toBe(true);
        (0, vitest_1.expect)(validator.validate(validPacket({ confidence: 1 })).ok).toBe(true);
    });
    (0, vitest_1.it)("rejects unregistered target", () => {
        const result = validator.validate(validPacket({ target: "ghost" }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.code).toBe("UNKNOWN_TARGET");
    });
    (0, vitest_1.it)("rejects packet type not accepted by target", () => {
        const result = validator.validate(validPacket({ type: "review.request" }));
        (0, vitest_1.expect)(result.ok).toBe(false);
        if (!result.ok)
            (0, vitest_1.expect)(result.code).toBe("TARGET_REJECTS_TYPE");
    });
});
