// @marsulta/mailman/dist/packet/__tests__/createPacket.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const createPacket_1 = require("../createPacket");
(0, vitest_1.describe)("createPacket", () => {
    (0, vitest_1.it)("fills in packetId if not provided", () => {
        const pkt = (0, createPacket_1.createPacket)({
            taskId: "t1",
            type: "task.assign",
            sender: "orch",
            target: "worker",
            payload: {},
        });
        (0, vitest_1.expect)(pkt.packetId).toBeTruthy();
        (0, vitest_1.expect)(typeof pkt.packetId).toBe("string");
    });
    (0, vitest_1.it)("fills in timestamp if not provided", () => {
        const pkt = (0, createPacket_1.createPacket)({
            taskId: "t1",
            type: "task.assign",
            sender: "orch",
            target: "worker",
            payload: {},
        });
        (0, vitest_1.expect)(pkt.timestamp).toBeTruthy();
        (0, vitest_1.expect)(() => new Date(pkt.timestamp)).not.toThrow();
    });
    (0, vitest_1.it)("preserves provided packetId", () => {
        const pkt = (0, createPacket_1.createPacket)({
            packetId: "my-custom-id",
            taskId: "t1",
            type: "task.assign",
            sender: "orch",
            target: "worker",
            payload: {},
        });
        (0, vitest_1.expect)(pkt.packetId).toBe("my-custom-id");
    });
    (0, vitest_1.it)("generates unique packetIds for separate calls", () => {
        const a = (0, createPacket_1.createPacket)({ taskId: "t", type: "task.assign", sender: "s", target: "t2", payload: {} });
        const b = (0, createPacket_1.createPacket)({ taskId: "t", type: "task.assign", sender: "s", target: "t2", payload: {} });
        (0, vitest_1.expect)(a.packetId).not.toBe(b.packetId);
    });
});
(0, vitest_1.describe)("createReply", () => {
    const original = (0, createPacket_1.createPacket)({
        packetId: "original-id",
        taskId: "task-99",
        type: "task.assign",
        sender: "orch",
        target: "worker",
        payload: { prompt: "go" },
    });
    (0, vitest_1.it)("sets parentPacketId to the original packetId", () => {
        const reply = (0, createPacket_1.createReply)(original, "task.result", "worker", { done: true });
        (0, vitest_1.expect)(reply.parentPacketId).toBe("original-id");
    });
    (0, vitest_1.it)("inherits taskId from the original", () => {
        const reply = (0, createPacket_1.createReply)(original, "task.result", "worker", { done: true });
        (0, vitest_1.expect)(reply.taskId).toBe("task-99");
    });
    (0, vitest_1.it)("flips sender and target", () => {
        const reply = (0, createPacket_1.createReply)(original, "task.result", "worker", { done: true });
        (0, vitest_1.expect)(reply.sender).toBe("worker");
        (0, vitest_1.expect)(reply.target).toBe("orch");
    });
    (0, vitest_1.it)("applies overrides", () => {
        const reply = (0, createPacket_1.createReply)(original, "task.result", "worker", {}, { status: "completed" });
        (0, vitest_1.expect)(reply.status).toBe("completed");
    });
});
